// clai-helpers/dist/transformers/claude-to-gemini-root.js
/**
 * claude-to-gemini-root: Compose GEMINI.md root instruction file.
 *
 * Source match: CLAUDE.md (used as a trigger — the transformer runs once).
 *
 * Composition: GEMINI.md is NOT a literal copy of CLAUDE.md. It aggregates
 * stack-agnostic instruction sources into a single Gemini-root document:
 *
 *   # GEMINI Instructions
 *   ## CATCHPHRASES
 *   [body of .github/instructions/persona/phrases/copilot-instructions.md]
 *   ## AI Coding Standards & Engineering Guide
 *   [body of .github/instructions/coding/copilot-instructions.md]
 *
 * Sources intentionally EXCLUDED:
 *   - .github/instructions/project/copilot-instructions.md
 *       — project overview is upstream-specific (UnderUndre/ai) and would
 *         confuse downstream consumers who have their own project identity.
 *   - CLAUDE.md body itself
 *       — CLAUDE.md is Claude Code-specific (router with tool-specific tips
 *         like context7, git MCP, /compact). Gemini doesn't need it.
 *
 * Consumer customization: add project-specific content via Protected Slots
 * (`<!-- HELPERS:CUSTOM START --> ... END -->`) inside GEMINI.md; those
 * survive every `sync`.
 */
import { FileKind } from "../types/common.js";
import { injectHeader } from "../core/header.js";
const PHRASES_PATH = ".github/instructions/persona/phrases/copilot-instructions.md";
const CODING_PATH = ".github/instructions/coding/copilot-instructions.md";
/**
 * Ordered composition. Order matters: catchphrases frame the voice, then
 * coding standards provide the rules. A consumer reading top-to-bottom gets
 * tone first, rules second.
 */
const SECTIONS = [
    { title: "## CATCHPHRASES", sourcePath: PHRASES_PATH },
    { title: "## AI Coding Standards & Engineering Guide", sourcePath: CODING_PATH },
];
const claudeToGeminiRoot = (_source, ctx) => {
    const parsed = ctx.allParsedFiles ?? [];
    const parts = ["# GEMINI Instructions", ""];
    for (const section of SECTIONS) {
        const file = parsed.find((p) => p.sourcePath === section.sourcePath);
        if (!file) {
            // Graceful degradation: missing source = skip that section. Do not
            // fail the whole run. A consumer running `init` against a mirror that
            // omits one of these paths still gets a valid partial GEMINI.md.
            continue;
        }
        parts.push(section.title, "");
        const stripped = stripFirstHeading(file.body);
        // Demote inner headings one level so the embedded doc's H2/H3/...
        // slot under our H2 section header consistently. Without this, inner
        // `## Subsection` rendered at the same outline level as the outer
        // `## CATCHPHRASES`, flattening the hierarchy.
        const demoted = demoteHeadings(stripped);
        parts.push(demoted.trim());
        parts.push(""); // blank line between sections
    }
    let body = parts.join("\n");
    // Rewrite cross-references from Claude paths to Gemini equivalents so
    // links inside GEMINI.md point to Gemini-format files, not Claude sources.
    body = body.replace(/\.claude\/agents\/(\w+)\.md/g, ".gemini/agents/$1.md");
    body = body.replace(/\.claude\/commands\/(\w+)\.md/g, ".gemini/commands/$1.toml");
    // Strip Protected Slot markers and their content from the composed body.
    // Slots are a .claude/ authoring convention and have no meaning in the
    // Gemini output (slots survive sync by being injected during apply, not
    // from source text).
    body = body.replace(/<!--\s*HELPERS:CUSTOM\s+START\s*-->[\s\S]*?<!--\s*HELPERS:CUSTOM\s+END\s*-->\n?/g, "");
    // Collapse runs of blank lines (artifacts of section stitching).
    body = body.replace(/\n{3,}/g, "\n\n").trim();
    const content = injectHeader(`\n${body}\n`, "CLAUDE.md", ".md");
    return {
        targetPath: "GEMINI.md",
        content,
        kind: FileKind.Generated,
        fromSource: "CLAUDE.md",
        transformer: "claude-to-gemini-root",
    };
};
/**
 * Demote every ATX heading in `body` by one level: `##` → `###`, `###` →
 * `####`, etc. H1 (`#`) is also demoted to `##`, but the first H1 is
 * expected to have been stripped by `stripFirstHeading` already. Caps at
 * level 6 (Markdown limit).
 *
 * Detection is conservative: only lines whose first non-whitespace token
 * matches `#{1,6}` followed by a space are treated as headings. Code blocks
 * (fenced) are skipped so example headings inside triple-backticks don't
 * get rewritten.
 */
function demoteHeadings(body) {
    const lines = body.split(/\r?\n/);
    const out = [];
    let inFence = false;
    let fenceMarker = "";
    for (const line of lines) {
        // Track fenced code blocks — don't rewrite inside them.
        const fenceMatch = line.match(/^(\s*)(```+|~~~+)/);
        if (fenceMatch) {
            const marker = fenceMatch[2];
            if (!inFence) {
                inFence = true;
                fenceMarker = marker;
            }
            else if (marker.startsWith(fenceMarker[0]) && marker.length >= fenceMarker.length) {
                inFence = false;
                fenceMarker = "";
            }
            out.push(line);
            continue;
        }
        if (!inFence) {
            const headingMatch = line.match(/^(#{1,6})(\s+.*)$/);
            if (headingMatch) {
                const level = headingMatch[1].length;
                const rest = headingMatch[2];
                const newLevel = Math.min(level + 1, 6);
                out.push("#".repeat(newLevel) + rest);
                continue;
            }
        }
        out.push(line);
    }
    return out.join("\n");
}
/**
 * Drop the first H1 (`# Title`) from a parsed body so that when we inject
 * the body under our own `## Section` header, we don't end up with an
 * orphan H1 floating inside a larger document.
 */
function stripFirstHeading(body) {
    const lines = body.split(/\r?\n/);
    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        if (line.trim() === "")
            continue;
        if (line.startsWith("# ")) {
            return lines.slice(i + 1).join("\n");
        }
        break; // first non-empty line isn't an H1 — keep body unchanged
    }
    return body;
}
export default claudeToGeminiRoot;
