// clai-helpers/dist/core/fleet/config.js
/**
 * Fleet config loader.
 * Loads ~/.config/clai-helpers/fleet.{json,ts,js,yaml,yml} via c12.
 * Contract: specs/003-fleet-sync/contracts/config-schema.md
 */
import { existsSync } from "node:fs";
import { writeFile } from "node:fs/promises";
import { join } from "node:path";
import { homedir } from "node:os";
import { loadConfig } from "c12";
import { createConsola } from "consola";
import { FleetError } from "./types.js";
const log = createConsola().withTag("fleet");
// ── Constants ───────────────────────────────────────────────────────
const VALID_SYNC_MODES = ["pr", "push", "patch"];
const ORG_NAME_RE = /^[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?$/;
const CONFIG_DIR_NAME = "clai-helpers";
const CONFIG_BASENAME = "fleet";
const DEFAULT_CONFIG = {
    scope: {
        includeOwnRepos: true,
        orgs: [],
        filter: undefined,
    },
    defaultSyncMode: "pr",
    patchOutputDir: "./.fleet-patches",
    discoveryConcurrency: 5,
};
// ── Path resolution ─────────────────────────────────────────────────
function configDir() {
    // Windows: %APPDATA%, Linux/macOS: XDG_CONFIG_HOME or ~/.config
    const appData = process.env.APPDATA;
    if (appData)
        return join(appData, CONFIG_DIR_NAME);
    const xdg = process.env.XDG_CONFIG_HOME;
    if (xdg)
        return join(xdg, CONFIG_DIR_NAME);
    return join(homedir(), ".config", CONFIG_DIR_NAME);
}
function configPaths() {
    const dir = configDir();
    return {
        dir,
        jsonPath: join(dir, `${CONFIG_BASENAME}.json`),
        tsPath: join(dir, `${CONFIG_BASENAME}.config.ts`),
    };
}
// ── Validation ──────────────────────────────────────────────────────
function validateOrgName(org) {
    if (!ORG_NAME_RE.test(org)) {
        throw new FleetError("config/invalid-scope", `Invalid org name: "${org}". Must match ${ORG_NAME_RE.source}`);
    }
}
function validateConfig(raw) {
    // Start from defaults
    const config = { ...DEFAULT_CONFIG, scope: { ...DEFAULT_CONFIG.scope } };
    // Validate scope
    if (raw.scope && typeof raw.scope === "object") {
        const scope = raw.scope;
        if ("includeOwnRepos" in scope) {
            if (typeof scope.includeOwnRepos !== "boolean") {
                throw new FleetError("config/malformed", `scope.includeOwnRepos must be boolean, got ${typeof scope.includeOwnRepos}`);
            }
            config.scope.includeOwnRepos = scope.includeOwnRepos;
        }
        if ("orgs" in scope) {
            if (!Array.isArray(scope.orgs)) {
                throw new FleetError("config/malformed", `scope.orgs must be an array, got ${typeof scope.orgs}`);
            }
            for (const org of scope.orgs) {
                if (typeof org !== "string") {
                    throw new FleetError("config/malformed", `scope.orgs must contain strings, got ${typeof org}`);
                }
                validateOrgName(org);
            }
            config.scope.orgs = scope.orgs;
        }
        if ("filter" in scope && scope.filter !== null && scope.filter !== undefined) {
            if (typeof scope.filter !== "string") {
                throw new FleetError("config/malformed", `scope.filter must be a string or null, got ${typeof scope.filter}`);
            }
            config.scope.filter = scope.filter;
        }
    }
    // Validate defaultSyncMode
    if ("defaultSyncMode" in raw) {
        if (typeof raw.defaultSyncMode !== "string" || !VALID_SYNC_MODES.includes(raw.defaultSyncMode)) {
            throw new FleetError("config/malformed", `defaultSyncMode must be one of ${VALID_SYNC_MODES.join(", ")}, got "${raw.defaultSyncMode}"`);
        }
        config.defaultSyncMode = raw.defaultSyncMode;
    }
    // Validate patchOutputDir
    if ("patchOutputDir" in raw) {
        if (typeof raw.patchOutputDir !== "string") {
            throw new FleetError("config/malformed", `patchOutputDir must be a string, got ${typeof raw.patchOutputDir}`);
        }
        config.patchOutputDir = raw.patchOutputDir;
    }
    // Validate discoveryConcurrency
    if ("discoveryConcurrency" in raw) {
        const val = raw.discoveryConcurrency;
        if (typeof val !== "number" || !Number.isInteger(val) || val < 1 || val > 20) {
            throw new FleetError("config/malformed", `discoveryConcurrency must be an integer in [1, 20], got ${val}`);
        }
        config.discoveryConcurrency = val;
    }
    // Warn on unknown keys
    const knownKeys = new Set(["scope", "defaultSyncMode", "patchOutputDir", "discoveryConcurrency"]);
    for (const key of Object.keys(raw)) {
        if (!knownKeys.has(key)) {
            log.warn(`unknown config key '${key}'`);
        }
    }
    return config;
}
// ── Public API ──────────────────────────────────────────────────────
/**
 * Load fleet config from ~/.config/clai-helpers/fleet.{json,ts,js,yaml,yml}.
 * Returns defaults when no config file exists.
 */
export async function loadFleetConfig() {
    const { dir } = configPaths();
    const { config: raw } = await loadConfig({
        cwd: dir,
        name: CONFIG_BASENAME,
        defaults: DEFAULT_CONFIG,
    });
    // c12 returns the merged config; validate it
    return validateConfig(raw ?? {});
}
/**
 * Add an org to scope.orgs. Writes back as JSON.
 * Idempotent: if org already present, no change.
 */
export async function addOrg(org) {
    validateOrgName(org);
    const paths = configPaths();
    // Refuse TS config mutation
    if (existsSync(paths.tsPath)) {
        throw new FleetError("config/malformed", "config is in TypeScript form; edit manually or convert to JSON for automated mutation");
    }
    const config = await loadFleetConfig();
    if (config.scope.orgs.includes(org)) {
        // Idempotent: already present
        log.info(`org "${org}" already in config`);
        return;
    }
    const updated = {
        ...config,
        scope: {
            ...config.scope,
            orgs: [...config.scope.orgs, org].sort(),
        },
    };
    await writeJsonConfig(paths.jsonPath, updated);
    log.info(`added org "${org}" to fleet config`);
}
/**
 * Remove an org from scope.orgs. Writes back as JSON.
 * Idempotent: if org not present, no change.
 */
export async function removeOrg(org) {
    validateOrgName(org);
    const paths = configPaths();
    // Refuse TS config mutation
    if (existsSync(paths.tsPath)) {
        throw new FleetError("config/malformed", "config is in TypeScript form; edit manually or convert to JSON for automated mutation");
    }
    const config = await loadFleetConfig();
    if (!config.scope.orgs.includes(org)) {
        // Idempotent: not present
        log.info(`org "${org}" not in config`);
        return;
    }
    const updated = {
        ...config,
        scope: {
            ...config.scope,
            orgs: config.scope.orgs.filter((o) => o !== org),
        },
    };
    await writeJsonConfig(paths.jsonPath, updated);
    log.info(`removed org "${org}" from fleet config`);
}
// ── Internal helpers ────────────────────────────────────────────────
async function writeJsonConfig(path, config) {
    const { mkdir } = await import("node:fs/promises");
    const { dirname } = await import("node:path");
    // Ensure directory exists
    await mkdir(dirname(path), { recursive: true });
    const json = JSON.stringify(config, null, 2) + "\n";
    await writeFile(path, json, "utf-8");
}
/**
 * Get the config directory path (exported for testing).
 */
export function getConfigDir() {
    return configDir();
}
/**
 * Get config file paths (exported for testing).
 */
export function getConfigPaths() {
    return configPaths();
}
