// clai-helpers/dist/core/fleet/github-api.js
/**
 * GitHub REST API v3 wrapper for Fleet Sync.
 * Contract: specs/003-fleet-sync/contracts/github-api-surface.md
 */
import { resolveAuth } from "../fetch.js";
import { FleetError } from "./types.js";
import { createConsola } from "consola";
const log = createConsola().withTag("fleet");
const GITHUB_API = "https://api.github.com";
function headers(auth) {
    return {
        Accept: "application/vnd.github+json",
        Authorization: `Bearer ${auth}`,
        "User-Agent": "clai-helpers",
    };
}
/**
 * Sleep for `ms` milliseconds.
 */
function sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}
/**
 * Extract the next page URL from the Link header.
 */
function parseNextLink(linkHeader) {
    if (!linkHeader)
        return undefined;
    const match = linkHeader.match(/<([^>]+)>;\s*rel="next"/);
    return match?.[1];
}
/**
 * Core request wrapper with error mapping, rate-limit handling, and retries.
 */
async function githubRequest(url, auth, fetchFn, method = "GET", body) {
    const maxRetries = 3;
    let lastError;
    for (let attempt = 0; attempt <= maxRetries; attempt++) {
        const response = await fetchFn(url, {
            method,
            headers: {
                ...headers(auth),
                ...(body !== undefined ? { "Content-Type": "application/json" } : {}),
            },
            body: body !== undefined ? JSON.stringify(body) : undefined,
        });
        // Rate-limit awareness: warn if remaining is low
        const remaining = response.headers.get("x-ratelimit-remaining");
        if (remaining !== null && Number(remaining) < 50) {
            log.warn(`Rate limit low: ${remaining} requests remaining`);
        }
        // Success
        if (response.ok)
            return response;
        // 401 → auth/missing
        if (response.status === 401) {
            throw new FleetError("auth/missing", "Authentication required. Set GH_TOKEN or run `gh auth login`.");
        }
        // 403 handling
        if (response.status === 403) {
            const isRateLimited = remaining === "0" ||
                response.headers.get("retry-after") !== null;
            if (isRateLimited) {
                const retryAfter = response.headers.get("retry-after");
                const resetHeader = response.headers.get("x-ratelimit-reset");
                let waitMs;
                if (retryAfter !== null) {
                    waitMs = Math.min(Number(retryAfter) * 1000, 60_000);
                }
                else if (resetHeader !== null) {
                    const resetAt = Number(resetHeader) * 1000;
                    waitMs = Math.min(resetAt - Date.now(), 60_000);
                }
                else {
                    waitMs = 2000 * Math.pow(2, attempt);
                }
                if (waitMs > 60_000 || attempt >= maxRetries) {
                    throw new FleetError("github/rate-limited", "GitHub API rate limit exceeded. Retry later.");
                }
                log.warn(`Rate limited. Waiting ${Math.round(waitMs / 1000)}s before retry (${attempt + 1}/${maxRetries})`);
                await sleep(waitMs);
                continue;
            }
            // 403 without rate-limit headers → insufficient scope
            throw new FleetError("auth/insufficient-scope", "Token lacks required scope for this operation.");
        }
        // 404
        if (response.status === 404) {
            throw new FleetError("github/repo-not-found", `Repository not found: ${url}`);
        }
        // 5xx → retry once then throw
        if (response.status >= 500) {
            if (attempt < 1) {
                const backoff = 2000 * Math.pow(2, attempt);
                log.warn(`Server error ${response.status}. Retrying in ${backoff}ms`);
                await sleep(backoff);
                continue;
            }
            throw new FleetError("github/network", `GitHub server error: ${response.status}`);
        }
        // Other errors
        throw new FleetError("github/api-error", `GitHub API error: ${response.status} ${response.statusText}`);
    }
    // Should not reach here, but just in case
    throw lastError ?? new FleetError("github/api-error", "Unexpected error in GitHub API request");
}
/**
 * Paginate through GitHub list endpoints following Link rel="next".
 */
async function paginate(url, auth, fetchFn) {
    const results = [];
    let nextUrl = url;
    while (nextUrl) {
        const response = await githubRequest(nextUrl, auth, fetchFn);
        const page = (await response.json());
        results.push(...page);
        const linkHeader = response.headers.get("link");
        nextUrl = parseNextLink(linkHeader);
    }
    return results;
}
// ── Public API ──────────────────────────────────────────────────────
/**
 * Resolve auth token. Uses resolveAuth() from core/fetch.ts.
 */
export function resolveFleetAuth() {
    const token = resolveAuth();
    if (!token) {
        throw new FleetError("auth/missing", "No GitHub token found. Set GH_TOKEN or run `gh auth login`.");
    }
    return token;
}
/**
 * List repos owned by the authenticated user.
 * GET /user/repos?affiliation=owner&per_page=100
 */
export async function listReposForUser(auth, fetchFn = globalThis.fetch) {
    return paginate(`${GITHUB_API}/user/repos?affiliation=owner&per_page=100`, auth, fetchFn);
}
/**
 * List repos for an org.
 * GET /orgs/{org}/repos?type=all&per_page=100
 */
export async function listReposForOrg(org, auth, fetchFn = globalThis.fetch) {
    return paginate(`${GITHUB_API}/orgs/${org}/repos?type=all&per_page=100`, auth, fetchFn);
}
/**
 * Read helpers-lock.json from a repo, decode and parse.
 * GET /repos/{owner}/{repo}/contents/helpers-lock.json?ref={branch}
 */
export async function readLockfile(owner, repo, branch, auth, fetchFn = globalThis.fetch) {
    const url = `${GITHUB_API}/repos/${owner}/${repo}/contents/helpers-lock.json?ref=${branch}`;
    const response = await githubRequest(url, auth, fetchFn);
    const data = (await response.json());
    if (data.encoding !== "base64" || !data.content) {
        throw new FleetError("lockfile/malformed", `Unexpected encoding or empty content in ${owner}/${repo}`);
    }
    // GitHub Base64 includes newlines — strip them before decoding
    const decoded = atob(data.content.replace(/\n/g, ""));
    let parsed;
    try {
        parsed = JSON.parse(decoded);
    }
    catch {
        throw new FleetError("lockfile/malformed", `helpers-lock.json in ${owner}/${repo} is not valid JSON`);
    }
    if (typeof parsed !== "object" ||
        parsed === null ||
        !("ref" in parsed) ||
        !("source" in parsed) ||
        typeof parsed.ref !== "string" ||
        typeof parsed.source !== "string") {
        throw new FleetError("lockfile/malformed", `helpers-lock.json in ${owner}/${repo} missing 'ref' or 'source'`);
    }
    return {
        ref: parsed.ref,
        source: parsed.source,
    };
}
/**
 * Get the most recent commit date for a path.
 * GET /repos/{owner}/{repo}/commits?path={path}&per_page=1
 */
export async function readLastCommitForPath(owner, repo, path, auth, fetchFn = globalThis.fetch) {
    const url = `${GITHUB_API}/repos/${owner}/${repo}/commits?path=${encodeURIComponent(path)}&per_page=1`;
    const response = await githubRequest(url, auth, fetchFn);
    const commits = (await response.json());
    if (commits.length === 0)
        return null;
    return commits[0].commit.author.date;
}
/**
 * Get the default branch for a repo.
 * GET /repos/{owner}/{repo}
 */
export async function getDefaultBranch(owner, repo, auth, fetchFn = globalThis.fetch) {
    const url = `${GITHUB_API}/repos/${owner}/${repo}`;
    const response = await githubRequest(url, auth, fetchFn);
    const data = (await response.json());
    return data.default_branch;
}
/**
 * Get the latest release tag for a repo.
 * GET /repos/{owner}/{repo}/releases/latest
 */
export async function getLatestRelease(owner, repo, auth, fetchFn = globalThis.fetch) {
    const url = `${GITHUB_API}/repos/${owner}/${repo}/releases/latest`;
    const response = await githubRequest(url, auth, fetchFn);
    const data = (await response.json());
    return data.tag_name;
}
/**
 * Find an open pull request for a given branch.
 * GET /repos/{owner}/{repo}/pulls?head={owner}:{branchName}&state=open
 */
export async function findOpenPullRequest(owner, repo, branchName, auth, fetchFn = globalThis.fetch) {
    const url = `${GITHUB_API}/repos/${owner}/${repo}/pulls?head=${owner}:${encodeURIComponent(branchName)}&state=open`;
    const response = await githubRequest(url, auth, fetchFn);
    const pulls = (await response.json());
    if (pulls.length === 0)
        return null;
    return pulls[0];
}
/**
 * Create a pull request.
 * POST /repos/{owner}/{repo}/pulls
 */
export async function createPullRequest(owner, repo, params, auth, fetchFn = globalThis.fetch) {
    const url = `${GITHUB_API}/repos/${owner}/${repo}/pulls`;
    const response = await githubRequest(url, auth, fetchFn, "POST", params);
    return (await response.json());
}
