// clai-helpers/dist/cli/add-target.js
/**
 * `helpers add-target` command handler.
 * Enable a new target and generate its files.
 */
import { defineCommand } from "citty";
import consola from "consola";
import { readFile, readdir } from "node:fs/promises";
import { join, relative } from "pathe";
import { ExitCode, FileKind, FileStatus } from "../types/common.js";
import { readLock, writeLock } from "../core/lock.js";
import { fetchSource } from "../core/fetch.js";
import { loadManifest } from "../core/manifest.js";
import { preloadAllTransformers, getTransformer } from "../transformers/registry.js";
import { parseSourceFile } from "../core/parse.js";
import { matchGlob, resolveOutputPath } from "../core/glob.js";
import { renderedHash } from "../core/hash.js";
import { createStagingDir, stageFile, commitStaged, cleanStaging, } from "../core/staging.js";
import { guardMutatingCommand, releaseMutatingGuard } from "../cli.js";
export default defineCommand({
    meta: {
        name: "add-target",
        description: "Enable a new target and generate its files",
    },
    args: {
        name: {
            type: "positional",
            description: "Target name to add",
            required: true,
        },
    },
    async run({ args }) {
        const root = process.cwd();
        let lockAcquired = false;
        try {
            lockAcquired = await guardMutatingCommand(root, "add-target");
            const lock = await readLock(root);
            if (!lock) {
                consola.error("No helpers-lock.json found. Run `helpers init` first.");
                process.exitCode = ExitCode.UsageError;
                return;
            }
            const targetName = args.name;
            if (lock.targets.includes(targetName)) {
                consola.error(`Target "${targetName}" is already active.`);
                process.exitCode = ExitCode.UsageError;
                return;
            }
            // Fetch source
            const { dir: sourceDir } = await fetchSource(lock.source.url, lock.source.ref, undefined, { preferOffline: true });
            const manifest = await loadManifest(sourceDir);
            const target = manifest.targets[targetName];
            if (!target) {
                consola.error(`Target "${targetName}" not found in manifest. Available: ${Object.keys(manifest.targets).join(", ")}`);
                process.exitCode = ExitCode.UsageError;
                return;
            }
            await preloadAllTransformers(manifest, [targetName]);
            // Discover and parse source files
            const sourceFiles = await discoverFiles(sourceDir, manifest.sources);
            const parsedFiles = await Promise.all(sourceFiles.map(async (relPath) => {
                const content = await readFile(join(sourceDir, relPath), "utf8");
                return parseSourceFile(relPath, content);
            }));
            // Run pipelines
            const stagingDir = await createStagingDir(root);
            const ctx = { sourceCommit: lock.source.commit, toolVersion: lock.toolVersion, targetName, config: manifest };
            for (const pipeline of target.pipelines) {
                const transformer = getTransformer(pipeline.transformer);
                for (const parsed of parsedFiles) {
                    if (!matchGlob(pipeline.match, parsed.sourcePath))
                        continue;
                    const result = transformer(parsed, ctx);
                    if (!result)
                        continue;
                    const files = Array.isArray(result) ? result : [result];
                    for (const file of files) {
                        file.targetPath = resolveOutputPath(pipeline.output, parsed.sourcePath, pipeline.match);
                        const stagedPath = await stageFile(stagingDir, file.content, file.targetPath);
                        await commitStaged(stagedPath, join(root, file.targetPath));
                        // Build source path mapping
                        const sourcePathMap = new Map();
                        for (const entry of lock.files) {
                            if (entry.kind === "source") {
                                const sourceRel = entry.path.replace(/^\.claude\//, "");
                                sourcePathMap.set(sourceRel, entry.path);
                                if (entry.path === sourceRel)
                                    sourcePathMap.set(entry.path, entry.path);
                            }
                        }
                        lock.files.push({
                            path: file.targetPath,
                            kind: FileKind.Generated,
                            transformer: file.transformer,
                            fromSource: sourcePathMap.get(file.fromSource) ?? file.fromSource,
                            renderedHash: renderedHash(file.content),
                            localRenderedHash: renderedHash(file.content),
                            status: FileStatus.Managed,
                        });
                    }
                }
            }
            lock.targets.push(targetName);
            await writeLock(root, lock);
            await cleanStaging(root);
            consola.success(`Target "${targetName}" added.`);
        }
        finally {
            await releaseMutatingGuard(root, lockAcquired);
        }
    },
});
async function discoverFiles(baseDir, patterns) {
    const files = [];
    async function walk(dir) {
        let entries;
        try {
            entries = await readdir(dir, { withFileTypes: true });
        }
        catch (e) {
            consola.warn(`Could not read directory ${dir}:`, e);
            return;
        }
        for (const entry of entries) {
            if (entry.isDirectory())
                await walk(join(dir, entry.name));
            else {
                const rel = relative(baseDir, join(dir, entry.name));
                if (patterns.some((p) => matchGlob(p, rel)))
                    files.push(rel);
            }
        }
    }
    await walk(baseDir);
    return files.sort();
}
