// clai-helpers/dist/cli/diff.js
/**
 * `helpers diff` command handler.
 * Show what would change on next sync.
 */
import { defineCommand } from "citty";
import consola from "consola";
import { ExitCode } from "../types/common.js";
import { readLock } from "../core/lock.js";
import { fetchSource } from "../core/fetch.js";
import { computeSyncPlan } from "./sync.js";
export default defineCommand({
    meta: {
        name: "diff",
        description: "Show what would change on next sync",
    },
    args: {
        targets: {
            type: "string",
            description: "Filter to specific targets",
        },
    },
    async run({ args }) {
        const root = process.cwd();
        const offline = args.offline === true;
        const lock = await readLock(root);
        if (!lock) {
            consola.error("No helpers-lock.json found. Run `helpers init` first.");
            process.exitCode = ExitCode.UsageError;
            return;
        }
        const { dir: sourceDir } = await fetchSource(lock.source.url, lock.source.ref, undefined, { offline, preferOffline: true });
        const plan = await computeSyncPlan(lock, sourceDir, lock.targets, {});
        const changes = plan.items.filter((i) => i.action !== "skip");
        if (changes.length === 0) {
            consola.success("No changes — local files match source.");
            return;
        }
        for (const item of plan.items) {
            const symbol = item.action === "add" ? "+" : item.action === "delete" ? "-" : item.action === "update" ? "~" : " ";
            consola.log(`  ${symbol} ${item.path} — ${item.reason}`);
        }
    },
});
