// clai-helpers/dist/cli/eject.js
/**
 * `helpers eject` command handler.
 * Untrack a file but keep it locally.
 */
import { defineCommand } from "citty";
import consola from "consola";
import { normalize } from "pathe";
import { ExitCode, FileStatus, FileKind } from "../types/common.js";
import { readLock, writeLock } from "../core/lock.js";
import { guardMutatingCommand, releaseMutatingGuard } from "../cli.js";
export default defineCommand({
    meta: {
        name: "eject",
        description: "Untrack a file but keep it locally",
    },
    args: {
        path: {
            type: "positional",
            description: "File path to eject",
            required: true,
        },
        cascade: {
            type: "boolean",
            default: false,
            description: "Also untrack generated descendants",
        },
    },
    async run({ args }) {
        const root = process.cwd();
        let lockAcquired = false;
        try {
            lockAcquired = await guardMutatingCommand(root, "eject");
            const lock = await readLock(root);
            if (!lock) {
                consola.error("No helpers-lock.json found.");
                process.exitCode = ExitCode.UsageError;
                return;
            }
            const targetPath = normalize(args.path);
            const entry = lock.files.find((e) => normalize(e.path) === targetPath);
            if (!entry) {
                consola.error(`File "${targetPath}" is not tracked.`);
                process.exitCode = ExitCode.UsageError;
                return;
            }
            entry.status = FileStatus.Ejected;
            let ejectedCount = 1;
            if (args.cascade && entry.kind === FileKind.Source) {
                for (const gen of lock.files) {
                    if (gen.kind === FileKind.Generated && gen.fromSource === entry.path) {
                        gen.status = FileStatus.Ejected;
                        ejectedCount++;
                    }
                }
            }
            await writeLock(root, lock);
            consola.success(`Ejected ${ejectedCount} file(s). They remain on disk but won't be managed by sync.`);
        }
        finally {
            await releaseMutatingGuard(root, lockAcquired);
        }
    },
});
