// clai-helpers/dist/cli/init.js
/**
 * `helpers init` command handler.
 * Bootstrap a new project from source repo.
 */
import { defineCommand } from "citty";
import consola from "consola";
import { readdir, readFile, access } from "node:fs/promises";
import { join, relative } from "pathe";
import { ExitCode, FileKind, FileClass, FileStatus } from "../types/common.js";
import { fetchSource } from "../core/fetch.js";
import { loadManifest } from "../core/manifest.js";
import { readLock, writeLock } from "../core/lock.js";
import { preloadAllTransformers, getTransformer } from "../transformers/registry.js";
import { parseSourceFile } from "../core/parse.js";
import { matchGlob, resolveOutputPath } from "../core/glob.js";
import { canonicalHash, slotsHash, renderedHash } from "../core/hash.js";
import { parseSlots, mergeSlots } from "../core/slots.js";
import { createStagingDir, stageFile, commitStaged, cleanStaging, } from "../core/staging.js";
import { createJournal, writeJournal, markOperationDone, deleteJournal, } from "../core/journal.js";
import { ensureGitignoreEntries } from "../core/gitignore.js";
import { guardMutatingCommand, releaseMutatingGuard } from "../cli.js";
export default defineCommand({
    meta: {
        name: "init",
        description: "Bootstrap a new project from source repo",
    },
    args: {
        source: {
            type: "string",
            default: "github:UnderUndre/ai",
            description: "Source repo URL",
        },
        version: {
            type: "string",
            description: "Pin to specific version tag",
        },
        ref: {
            type: "string",
            description: "Branch or SHA (overrides --version)",
        },
        targets: {
            type: "string",
            default: "claude,copilot,gemini",
            description: "Comma-separated target names",
        },
        "source-config": {
            type: "string",
            description: "Local manifest override",
        },
        "trust-custom": {
            type: "boolean",
            default: false,
            description: "Pre-approve custom transformers",
        },
    },
    async run({ args }) {
        const root = process.cwd();
        const dryRun = args["dry-run"] === true;
        let lockAcquired = false;
        try {
            if (!dryRun) {
                lockAcquired = await guardMutatingCommand(root, "init");
            }
            // Check if already initialized
            const existingLock = await readLock(root);
            if (existingLock) {
                consola.error("Project already initialized. Use `helpers sync` to update.");
                process.exitCode = ExitCode.UsageError;
                return;
            }
            // 1. Fetch source repo
            const ref = args.ref ?? args.version;
            consola.start("Fetching source repo...");
            const { dir: sourceDir, commit } = await fetchSource(args.source, ref, undefined, { offline: args.offline === true });
            consola.success(`Fetched source (commit: ${commit.slice(0, 8)})`);
            // 2. Load and validate manifest
            const manifest = await loadManifest(sourceDir, args["source-config"]);
            consola.success(`Loaded manifest: ${manifest.sources.length} source patterns, ${Object.keys(manifest.targets).length} targets`);
            // 3. Resolve active targets
            const requestedTargets = args.targets.split(",").map((t) => t.trim());
            for (const t of requestedTargets) {
                if (!manifest.targets[t]) {
                    consola.error(`Unknown target: "${t}". Available: ${Object.keys(manifest.targets).join(", ")}`);
                    process.exitCode = ExitCode.UsageError;
                    return;
                }
            }
            // 4. Preload all transformers for active targets
            await preloadAllTransformers(manifest, requestedTargets);
            // 5. Discover source files (from repo root, not .claude/ hardcoded)
            const sourceFiles = await discoverSourceFiles(sourceDir, manifest.sources);
            consola.info(`Found ${sourceFiles.length} source files`);
            // 6. Parse source files
            const parsedFiles = await Promise.all(sourceFiles.map(async (relPath) => {
                const content = await readFile(join(sourceDir, relPath), "utf8");
                return parseSourceFile(relPath, content);
            }));
            // 7. Run transformer pipelines per target
            const allRendered = [];
            const ctx = {
                sourceCommit: commit,
                toolVersion: "0.1.0",
                targetName: "",
                config: manifest,
                allParsedFiles: parsedFiles,
            };
            for (const targetName of requestedTargets) {
                const target = manifest.targets[targetName];
                ctx.targetName = targetName;
                for (const pipeline of target.pipelines) {
                    const transformer = getTransformer(pipeline.transformer);
                    for (const parsed of parsedFiles) {
                        if (!matchGlob(pipeline.match, parsed.sourcePath))
                            continue;
                        const result = transformer(parsed, ctx);
                        if (!result)
                            continue;
                        const files = Array.isArray(result) ? result : [result];
                        for (const file of files) {
                            // Always resolve output path from pipeline template
                            file.targetPath = resolveOutputPath(pipeline.output, parsed.sourcePath, pipeline.match);
                            allRendered.push(file);
                        }
                    }
                }
            }
            consola.info(`Generated ${allRendered.length} files across ${requestedTargets.length} targets`);
            // 8. Dry run — just print what would happen
            if (dryRun) {
                consola.box("DRY RUN — no files written");
                for (const file of allRendered) {
                    consola.log(`  ${file.targetPath} (from: ${file.fromSource}, via: ${file.transformer})`);
                }
                return;
            }
            // 9. Stage and write files via journal
            const stagingDir = await createStagingDir(root);
            const operations = allRendered.map((file, idx) => ({
                id: idx,
                op: "write",
                path: file.targetPath,
            }));
            const journal = createJournal(root, "init", operations, "");
            await writeJournal(root, journal);
            for (const file of allRendered) {
                // Check for existing file conflict
                try {
                    await access(join(root, file.targetPath));
                    // File exists — check if source with slots for merge
                    if (file.transformer === "identity") {
                        const existing = await readFile(join(root, file.targetPath), "utf8");
                        const ext = file.targetPath.match(/\.[^.]+$/)?.[0] ?? ".md";
                        const existingSlots = parseSlots(existing, ext);
                        if (existingSlots.length > 0) {
                            file.content = mergeSlots(file.content, existingSlots, ext);
                        }
                    }
                }
                catch {
                    // File doesn't exist — no conflict
                }
                const idx = allRendered.indexOf(file);
                const stagedPath = await stageFile(stagingDir, file.content, file.targetPath);
                // Write stagedPath to journal so recover --resume can find it
                const journalOp = journal.operations.find((op) => op.id === idx);
                if (journalOp) {
                    journalOp.stagedPath = stagedPath;
                    await writeJournal(root, journal);
                }
                await commitStaged(stagedPath, join(root, file.targetPath));
                await markOperationDone(root, journal, idx);
            }
            // 10. Build lock file
            const lock = {
                schema: 1,
                toolVersion: "0.1.0",
                source: {
                    url: args.source,
                    ref: ref ?? "latest",
                    commit,
                },
                installedAt: new Date().toISOString(),
                targets: requestedTargets,
                trustedTransformers: [],
                files: buildLockEntries(allRendered, parsedFiles, manifest),
            };
            await writeLock(root, lock);
            await deleteJournal(root);
            await cleanStaging(root);
            // Protect the consumer from accidentally committing the staging
            // directory the CLI uses for atomic writes. `helpers-lock.json` is
            // intentionally NOT added — it must be committed (same convention as
            // package-lock.json) so `status --strict` and team-wide drift checks work.
            const addedIgnores = await ensureGitignoreEntries(root, [
                { pattern: ".helpers/", comment: "staging directory for atomic writes" },
            ]);
            if (addedIgnores.length > 0) {
                consola.info(`Added to .gitignore: ${addedIgnores.join(", ")}`);
            }
            consola.success(`Initialized ${allRendered.length} files. Lock file written.`);
        }
        finally {
            await releaseMutatingGuard(root, lockAcquired);
        }
    },
});
/**
 * Recursively discover files matching source glob patterns under .claude/.
 */
async function discoverSourceFiles(baseDir, patterns) {
    const allFiles = [];
    async function walk(dir) {
        let entries;
        try {
            entries = await readdir(dir, { withFileTypes: true });
        }
        catch {
            return;
        }
        for (const entry of entries) {
            const fullPath = join(dir, entry.name);
            if (entry.isDirectory()) {
                await walk(fullPath);
            }
            else {
                const relPath = relative(baseDir, fullPath);
                // Check if matches any source pattern
                if (patterns.some((p) => matchGlob(p, relPath))) {
                    allFiles.push(relPath);
                }
            }
        }
    }
    await walk(baseDir);
    return allFiles.sort();
}
function buildLockEntries(rendered, _parsedFiles, _manifest) {
    const entries = [];
    const seen = new Set();
    // Build source-path mapping: sourcePath → lock entry path
    // (identity transformer creates the source entries)
    const sourcePathMap = new Map();
    for (const file of rendered) {
        if (file.transformer === "identity") {
            sourcePathMap.set(file.fromSource, file.targetPath);
        }
    }
    for (const file of rendered) {
        if (seen.has(file.targetPath))
            continue;
        seen.add(file.targetPath);
        if (file.transformer === "identity") {
            // Source entries (identity = source copy)
            const slots = parseSlots(file.content, file.targetPath.match(/\.[^.]+$/)?.[0] ?? ".md");
            const slotBodies = slots.map((s) => s.body);
            const entry = {
                path: file.targetPath,
                kind: FileKind.Source,
                class: FileClass.Core,
                status: FileStatus.Managed,
                sourceCanonicalHash: canonicalHash(file.content, slotBodies),
                localCanonicalHash: canonicalHash(file.content, slotBodies),
                slotsHash: slotsHash(slotBodies),
            };
            entries.push(entry);
        }
        else {
            // Generated entries — resolve fromSource to lock entry path
            const resolvedFromSource = sourcePathMap.get(file.fromSource) ?? file.fromSource;
            const entry = {
                path: file.targetPath,
                kind: FileKind.Generated,
                transformer: file.transformer,
                fromSource: resolvedFromSource,
                renderedHash: renderedHash(file.content),
                localRenderedHash: renderedHash(file.content),
                status: FileStatus.Managed,
            };
            entries.push(entry);
        }
    }
    return entries;
}
