// clai-helpers/dist/cli/recover.js
/**
 * `helpers recover` command handler.
 * Recover from a crashed sync/init run via journal.
 */
import { defineCommand } from "citty";
import consola from "consola";
import { copyFile, unlink, rm } from "node:fs/promises";
import { join } from "pathe";
import { ExitCode } from "../types/common.js";
import { readJournal, markOperationDone, deleteJournal, } from "../core/journal.js";
import { commitStaged, cleanStaging } from "../core/staging.js";
import { acquireProcessLock, releaseProcessLock } from "../core/process-lock.js";
export default defineCommand({
    meta: {
        name: "recover",
        description: "Recover from a crashed sync/init run",
    },
    args: {
        resume: {
            type: "boolean",
            default: false,
            description: "Re-attempt plan from first non-done entry",
        },
        rollback: {
            type: "boolean",
            default: false,
            description: "Restore from backup, return to pre-sync state",
        },
        abandon: {
            type: "boolean",
            default: false,
            description: "Delete journal + backup, leave files as-is (destructive)",
        },
    },
    async run({ args }) {
        const root = process.cwd();
        const yes = args.yes === true;
        const flagCount = [args.resume, args.rollback, args.abandon].filter(Boolean).length;
        if (flagCount !== 1) {
            consola.error("Exactly one of --resume, --rollback, or --abandon is required.");
            process.exitCode = ExitCode.UsageError;
            return;
        }
        // Acquire lock before touching any state.
        // acquireProcessLock auto-cleans stale PIDs; if it throws, a live process holds the lock.
        const lockAcquired = await acquireProcessLock(root);
        const journal = await readJournal(root);
        if (!journal) {
            consola.info("No stale journal found. Nothing to recover.");
            if (lockAcquired)
                await releaseProcessLock(root);
            return;
        }
        const incompleteOps = journal.operations.filter((op) => !op.done);
        consola.info(`Found stale journal from "${journal.command}" run (${journal.runId}). ` +
            `${incompleteOps.length} incomplete operations.`);
        try {
            if (args.resume) {
                consola.start("Resuming incomplete operations...");
                for (const op of incompleteOps) {
                    if (op.op === "write" && op.stagedPath) {
                        await commitStaged(op.stagedPath, join(root, op.path));
                    }
                    else if (op.op === "delete") {
                        try {
                            await unlink(join(root, op.path));
                        }
                        catch { /* already gone */ }
                    }
                    await markOperationDone(root, journal, op.id);
                }
                await deleteJournal(root);
                await cleanStaging(root);
                consola.success("Resume complete.");
            }
            else if (args.rollback) {
                consola.start("Rolling back to pre-sync state...");
                for (const op of journal.operations) {
                    if (op.backupPath) {
                        try {
                            await copyFile(op.backupPath, join(root, op.path));
                        }
                        catch {
                            consola.warn(`Could not restore backup for ${op.path}`);
                        }
                    }
                    else if (op.op === "write" && op.done) {
                        try {
                            await unlink(join(root, op.path));
                        }
                        catch { /* already gone */ }
                    }
                }
                await deleteJournal(root);
                await cleanStaging(root);
                await rm(join(root, ".helpers/backup", journal.runId), { recursive: true, force: true });
                consola.success("Rollback complete.");
            }
            else if (args.abandon) {
                if (!yes) {
                    consola.error("--abandon is destructive. Use --yes to confirm.");
                    process.exitCode = ExitCode.UsageError;
                    return;
                }
                await deleteJournal(root);
                await cleanStaging(root);
                await rm(join(root, ".helpers/backup", journal.runId), { recursive: true, force: true });
                consola.success("Journal abandoned. Files left as-is.");
            }
        }
        finally {
            if (lockAcquired)
                await releaseProcessLock(root);
        }
    },
});
