// clai-helpers/dist/cli/regen.js
/**
 * `helpers regen` command handler.
 *
 * In-place pipeline execution: reads `helpers.config.ts` from cwd, discovers
 * source files from cwd, runs all transformer pipelines, and writes outputs
 * back into cwd. Unlike `init`, there is NO network fetch and NO lockfile —
 * cwd itself IS the source of truth.
 *
 * Intended for upstream template repos (like UnderUndre/ai) that maintain
 * generated outputs alongside sources and need to refresh them after editing
 * the `.claude/` tree. Downstream consumer repos should use `sync` instead.
 *
 * Safety:
 *   - Refuses to run if no `helpers.config.ts` is found in cwd.
 *   - Uses the same staging + journal infrastructure as init/sync — if the
 *     process is killed mid-write, `helpers recover` can resume or rollback.
 *   - Protected Slots (`<!-- HELPERS:CUSTOM -->`) in existing managed files
 *     are preserved during regeneration.
 */
import { defineCommand } from "citty";
import consola from "consola";
import { readdir, readFile, access } from "node:fs/promises";
import { join, relative } from "pathe";
import { ExitCode } from "../types/common.js";
import { loadManifest } from "../core/manifest.js";
import { preloadAllTransformers, getTransformer } from "../transformers/registry.js";
import { parseSourceFile } from "../core/parse.js";
import { matchGlob, resolveOutputPath } from "../core/glob.js";
import { parseSlots, mergeSlots } from "../core/slots.js";
import { createStagingDir, stageFile, commitStaged, cleanStaging, } from "../core/staging.js";
import { createJournal, writeJournal, markOperationDone, deleteJournal, } from "../core/journal.js";
import { guardMutatingCommand, releaseMutatingGuard } from "../cli.js";
export default defineCommand({
    meta: {
        name: "regen",
        description: "Regenerate downstream targets in-place from the local helpers.config.ts (upstream-template repos only — no network fetch, no lockfile)",
    },
    args: {
        targets: {
            type: "string",
            description: "Comma-separated target names to regenerate. Defaults to ALL targets defined in helpers.config.ts",
        },
        "source-config": {
            type: "string",
            description: "Path override for helpers.config.ts (default: cwd)",
        },
        "trust-custom": {
            type: "boolean",
            default: false,
            description: "Pre-approve custom transformers",
        },
    },
    async run({ args }) {
        const root = process.cwd();
        const dryRun = args["dry-run"] === true;
        let lockAcquired = false;
        try {
            if (!dryRun) {
                lockAcquired = await guardMutatingCommand(root, "regen");
            }
            // 1. Verify we're in an upstream-template repo (has helpers.config.ts).
            const manifestPath = args["source-config"] ?? join(root, "helpers.config.ts");
            try {
                await access(manifestPath);
            }
            catch {
                consola.error("No helpers.config.ts found in cwd. `regen` is for upstream-template repos " +
                    "that maintain generated outputs alongside sources. Downstream consumer " +
                    "projects should use `sync` or `init` instead.");
                process.exitCode = ExitCode.UsageError;
                return;
            }
            // 2. Load manifest from cwd directly (no giget fetch).
            const manifest = await loadManifest(root, args["source-config"]);
            const availableTargets = Object.keys(manifest.targets);
            consola.success(`Loaded manifest: ${manifest.sources.length} source patterns, ${availableTargets.length} targets (${availableTargets.join(", ")})`);
            // 3. Resolve active targets (default: ALL targets).
            const requestedTargets = args.targets
                ? args.targets.split(",").map((t) => t.trim())
                : availableTargets;
            for (const t of requestedTargets) {
                if (!manifest.targets[t]) {
                    consola.error(`Unknown target: "${t}". Available: ${availableTargets.join(", ")}`);
                    process.exitCode = ExitCode.UsageError;
                    return;
                }
            }
            // 4. Preload transformers for the selected targets.
            await preloadAllTransformers(manifest, requestedTargets);
            // 5. Discover source files from cwd.
            const sourceFiles = await discoverSourceFiles(root, manifest.sources);
            consola.info(`Found ${sourceFiles.length} source files`);
            // 6. Parse source files.
            const parsedFiles = await Promise.all(sourceFiles.map(async (relPath) => {
                const content = await readFile(join(root, relPath), "utf8");
                return parseSourceFile(relPath, content);
            }));
            // 7. Run transformer pipelines.
            const allRendered = [];
            const ctx = {
                sourceCommit: "local",
                toolVersion: "0.1.0",
                targetName: "",
                config: manifest,
                allParsedFiles: parsedFiles,
            };
            for (const targetName of requestedTargets) {
                const target = manifest.targets[targetName];
                ctx.targetName = targetName;
                for (const pipeline of target.pipelines) {
                    const transformer = getTransformer(pipeline.transformer);
                    for (const parsed of parsedFiles) {
                        if (!matchGlob(pipeline.match, parsed.sourcePath))
                            continue;
                        const result = transformer(parsed, ctx);
                        if (!result)
                            continue;
                        const files = Array.isArray(result) ? result : [result];
                        for (const file of files) {
                            file.targetPath = resolveOutputPath(pipeline.output, parsed.sourcePath, pipeline.match);
                            allRendered.push(file);
                        }
                    }
                }
            }
            consola.info(`Generated ${allRendered.length} files across ${requestedTargets.length} targets`);
            // 8. Dry run — preview only.
            if (dryRun) {
                consola.box("DRY RUN — no files written");
                for (const file of allRendered) {
                    consola.log(`  ${file.targetPath} (from: ${file.fromSource}, via: ${file.transformer})`);
                }
                return;
            }
            // 9. Stage and commit via journal (same atomicity as init/sync).
            const stagingDir = await createStagingDir(root);
            const operations = allRendered.map((file, idx) => ({
                id: idx,
                op: "write",
                path: file.targetPath,
            }));
            const journal = createJournal(root, "regen", operations, "");
            await writeJournal(root, journal);
            for (let i = 0; i < allRendered.length; i++) {
                const file = allRendered[i];
                // Preserve Protected Slot content if the target already exists.
                try {
                    await access(join(root, file.targetPath));
                    if (file.transformer === "identity") {
                        const existing = await readFile(join(root, file.targetPath), "utf8");
                        const ext = file.targetPath.match(/\.[^.]+$/)?.[0] ?? ".md";
                        const existingSlots = parseSlots(existing, ext);
                        if (existingSlots.length > 0) {
                            file.content = mergeSlots(file.content, existingSlots, ext);
                        }
                    }
                }
                catch {
                    // Target doesn't exist yet — no slots to preserve
                }
                const stagedPath = await stageFile(stagingDir, file.content, file.targetPath);
                const journalOp = journal.operations.find((op) => op.id === i);
                if (journalOp) {
                    journalOp.stagedPath = stagedPath;
                    await writeJournal(root, journal);
                }
                await commitStaged(stagedPath, join(root, file.targetPath));
                await markOperationDone(root, journal, i);
            }
            await deleteJournal(root);
            await cleanStaging(root);
            consola.success(`Regenerated ${allRendered.length} files across ${requestedTargets.length} target(s). No lockfile written (upstream-in-place).`);
        }
        finally {
            await releaseMutatingGuard(root, lockAcquired);
        }
    },
});
/**
 * Recursively discover files matching source glob patterns under `baseDir`.
 * Same logic as init.ts — duplicated to keep regen self-contained without
 * tempting a premature extraction.
 */
async function discoverSourceFiles(baseDir, patterns) {
    const allFiles = [];
    async function walk(dir) {
        let entries;
        try {
            entries = await readdir(dir, { withFileTypes: true });
        }
        catch {
            return;
        }
        for (const entry of entries) {
            const fullPath = join(dir, entry.name);
            if (entry.isDirectory()) {
                // Skip common output/VCS dirs we never want to treat as sources.
                if (shouldSkipDir(entry.name))
                    continue;
                await walk(fullPath);
            }
            else {
                const relPath = relative(baseDir, fullPath);
                if (patterns.some((p) => matchGlob(p, relPath))) {
                    allFiles.push(relPath);
                }
            }
        }
    }
    await walk(baseDir);
    return allFiles.sort();
}
/**
 * Skip paths that can't possibly be sources and would just slow the walk.
 * Kept minimal — the real filter is the glob match against `manifest.sources`.
 */
function shouldSkipDir(name) {
    return (name === "node_modules" ||
        name === ".git" ||
        name === "dist" ||
        name === "build" ||
        name === ".helpers");
}
