// clai-helpers/dist/cli/remove.js
/**
 * `helpers remove` command handler.
 * Delete a file from disk and untrack it. Destructive.
 */
import { defineCommand } from "citty";
import consola from "consola";
import { unlink } from "node:fs/promises";
import { join, normalize } from "pathe";
import { ExitCode } from "../types/common.js";
import { readLock, writeLock } from "../core/lock.js";
import { guardMutatingCommand, releaseMutatingGuard } from "../cli.js";
export default defineCommand({
    meta: {
        name: "remove",
        description: "Delete a file from disk and untrack it",
    },
    args: {
        path: {
            type: "positional",
            description: "File path to remove",
            required: true,
        },
    },
    async run({ args }) {
        const root = process.cwd();
        const yes = args.yes === true;
        const interactive = args.interactive === true;
        if (!yes && !interactive) {
            consola.error("--remove is destructive. Use --yes or --interactive to confirm.");
            process.exitCode = ExitCode.UsageError;
            return;
        }
        let lockAcquired = false;
        try {
            lockAcquired = await guardMutatingCommand(root, "remove");
            const lock = await readLock(root);
            if (!lock) {
                consola.error("No helpers-lock.json found.");
                process.exitCode = ExitCode.UsageError;
                return;
            }
            const targetPath = normalize(args.path);
            const idx = lock.files.findIndex((e) => normalize(e.path) === targetPath);
            if (idx < 0) {
                consola.error(`File "${targetPath}" is not tracked.`);
                process.exitCode = ExitCode.UsageError;
                return;
            }
            // Delete from disk
            try {
                await unlink(join(root, targetPath));
            }
            catch {
                consola.warn(`File not found on disk: ${targetPath}`);
            }
            // Remove from lock
            lock.files.splice(idx, 1);
            await writeLock(root, lock);
            consola.success(`Removed "${targetPath}" from disk and lock.`);
        }
        finally {
            await releaseMutatingGuard(root, lockAcquired);
        }
    },
});
