// clai-helpers/dist/cli/remove-target.js
/**
 * `helpers remove-target` command handler.
 * Delete all files for a target and untrack. Destructive.
 */
import { defineCommand } from "citty";
import consola from "consola";
import { unlink } from "node:fs/promises";
import { join } from "pathe";
import { ExitCode } from "../types/common.js";
import { readLock, writeLock } from "../core/lock.js";
import { guardMutatingCommand, releaseMutatingGuard } from "../cli.js";
export default defineCommand({
    meta: {
        name: "remove-target",
        description: "Delete all files for a target and untrack",
    },
    args: {
        name: {
            type: "positional",
            description: "Target name to remove",
            required: true,
        },
    },
    async run({ args }) {
        const root = process.cwd();
        const yes = args.yes === true;
        const interactive = args.interactive === true;
        if (!yes && !interactive) {
            consola.error("--remove-target is destructive. Use --yes or --interactive to confirm.");
            process.exitCode = ExitCode.UsageError;
            return;
        }
        let lockAcquired = false;
        try {
            lockAcquired = await guardMutatingCommand(root, "remove-target");
            const lock = await readLock(root);
            if (!lock) {
                consola.error("No helpers-lock.json found.");
                process.exitCode = ExitCode.UsageError;
                return;
            }
            const targetName = args.name;
            if (!lock.targets.includes(targetName)) {
                consola.error(`Target "${targetName}" is not active.`);
                process.exitCode = ExitCode.UsageError;
                return;
            }
            // Find all generated files belonging to this target
            // We identify them by checking which transformer produced them and if the
            // transformer belongs to the target's pipelines
            const filesToRemove = lock.files.filter((entry) => {
                if (entry.kind !== "generated")
                    return false;
                // Heuristic: check transformer name prefix for target association
                // This works for built-in transformers which have target-specific names
                const transformer = entry.transformer;
                if (targetName === "copilot")
                    return transformer.includes("copilot");
                if (targetName === "gemini")
                    return transformer.includes("gemini");
                if (targetName === "claude")
                    return transformer === "identity";
                return false;
            });
            // Delete files from disk
            let deleted = 0;
            for (const entry of filesToRemove) {
                try {
                    await unlink(join(root, entry.path));
                    deleted++;
                }
                catch {
                    // Already deleted
                }
            }
            // Remove from lock
            const removePaths = new Set(filesToRemove.map((e) => e.path));
            lock.files = lock.files.filter((e) => !removePaths.has(e.path));
            lock.targets = lock.targets.filter((t) => t !== targetName);
            await writeLock(root, lock);
            consola.success(`Removed target "${targetName}": ${deleted} files deleted.`);
        }
        finally {
            await releaseMutatingGuard(root, lockAcquired);
        }
    },
});
