// clai-helpers/dist/cli/status.js
/**
 * `helpers status` command handler.
 * Show current state of tracked files with drift detection.
 */
import { defineCommand } from "citty";
import consola from "consola";
import { readFile } from "node:fs/promises";
import { join } from "pathe";
import { ExitCode } from "../types/common.js";
import { readLock } from "../core/lock.js";
import { detectDrift } from "../core/drift.js";
export default defineCommand({
    meta: {
        name: "status",
        description: "Show current state of tracked files",
    },
    args: {
        strict: {
            type: "boolean",
            default: false,
            description: "Exit non-zero on drift",
        },
        targets: {
            type: "string",
            description: "Filter to specific targets",
        },
    },
    async run({ args }) {
        const root = process.cwd();
        const json = args.json === true;
        const lock = await readLock(root);
        if (!lock) {
            consola.error("No helpers-lock.json found. Run `helpers init` first.");
            process.exitCode = ExitCode.UsageError;
            return;
        }
        const entries = [];
        let hasDrift = false;
        for (const entry of lock.files) {
            let localContent;
            try {
                localContent = await readFile(join(root, entry.path), "utf8");
            }
            catch {
                entries.push({
                    path: entry.path,
                    kind: entry.kind,
                    class: entry.kind === "source" ? entry.class : undefined,
                    status: "MISSING",
                    drift: true,
                    customized: false,
                    reason: "File not found on disk",
                });
                hasDrift = true;
                continue;
            }
            const ext = entry.path.match(/\.[^.]+$/)?.[0] ?? ".md";
            const result = detectDrift(entry, localContent, ext);
            entries.push({
                path: entry.path,
                kind: entry.kind,
                class: entry.kind === "source" ? entry.class : undefined,
                status: entry.status,
                drift: result.isDrift,
                customized: result.isCustomized,
                reason: result.reason,
            });
            if (result.isDrift)
                hasDrift = true;
        }
        if (json) {
            console.log(JSON.stringify(entries, null, 2));
        }
        else {
            // Table output
            consola.log("");
            consola.log(`  Source: ${lock.source.url} @ ${lock.source.ref} (${lock.source.commit.slice(0, 8)})`);
            consola.log(`  Targets: ${lock.targets.join(", ")}`);
            consola.log(`  Files: ${lock.files.length}`);
            consola.log("");
            for (const entry of entries) {
                const driftFlag = entry.drift ? " [DRIFT]" : entry.customized ? " [CUSTOMIZED]" : "";
                consola.log(`  ${entry.status.padEnd(12)} ${entry.kind.padEnd(10)} ${entry.path}${driftFlag}`);
            }
            if (hasDrift) {
                consola.log("");
                consola.warn("Drift detected in managed files.");
            }
        }
        if (args.strict && hasDrift) {
            process.exitCode = ExitCode.DriftDetected;
        }
    },
});
