// clai-helpers/dist/cli/sync.js
/**
 * `helpers sync` command handler.
 * Update existing project from source repo.
 * Two phases: Plan (pure computation) then Apply (side effects).
 */
import { defineCommand } from "citty";
import consola from "consola";
import { readFile, unlink, access } from "node:fs/promises";
import { join } from "pathe";
import { ExitCode, FileKind, FileClass, FileStatus } from "../types/common.js";
import { fetchSource } from "../core/fetch.js";
import { loadManifest } from "../core/manifest.js";
import { readLock, writeLock } from "../core/lock.js";
import { preloadAllTransformers, getTransformer } from "../transformers/registry.js";
import { parseSourceFile } from "../core/parse.js";
import { matchGlob, resolveOutputPath } from "../core/glob.js";
import { canonicalHash, slotsHash, renderedHash } from "../core/hash.js";
import { parseSlots, mergeSlots } from "../core/slots.js";
import { createStagingDir, stageFile, commitStaged, cleanStaging, } from "../core/staging.js";
import { createJournal, writeJournal, markOperationDone, deleteJournal, createBackup, } from "../core/journal.js";
import { guardMutatingCommand, releaseMutatingGuard } from "../cli.js";
import { readdir } from "node:fs/promises";
import { relative } from "pathe";
// ─── Plan Phase ─────────────────────────────────
export async function computeSyncPlan(lock, sourceDir, targetNames, options) {
    const manifest = await loadManifest(sourceDir);
    await preloadAllTransformers(manifest, targetNames);
    // Discover source files (from repo root, not .claude/ hardcoded)
    const sourceFiles = await discoverSourceFiles(sourceDir, manifest.sources);
    // Parse source files
    const parsedFiles = await Promise.all(sourceFiles.map(async (relPath) => {
        const content = await readFile(join(sourceDir, relPath), "utf8");
        return parseSourceFile(relPath, content);
    }));
    // Run all transformer pipelines
    const allRendered = [];
    const ctx = {
        sourceCommit: lock.source.commit,
        toolVersion: lock.toolVersion,
        targetName: "",
        config: manifest,
        allParsedFiles: parsedFiles,
    };
    // Build source-path mapping from identity transformer
    const sourcePathMap = new Map();
    for (const targetName of targetNames) {
        const target = manifest.targets[targetName];
        if (!target)
            continue;
        ctx.targetName = targetName;
        for (const pipeline of target.pipelines) {
            const transformer = getTransformer(pipeline.transformer);
            for (const parsed of parsedFiles) {
                if (!matchGlob(pipeline.match, parsed.sourcePath))
                    continue;
                const result = transformer(parsed, ctx);
                if (!result)
                    continue;
                const files = Array.isArray(result) ? result : [result];
                for (const file of files) {
                    file.targetPath = resolveOutputPath(pipeline.output, parsed.sourcePath, pipeline.match);
                    if (pipeline.transformer === "identity") {
                        sourcePathMap.set(file.fromSource, file.targetPath);
                    }
                    allRendered.push(file);
                }
            }
        }
    }
    // Build plan by comparing rendered files against lock
    const items = [];
    const renderedPaths = new Set(allRendered.map((f) => f.targetPath));
    const seen = new Set();
    for (const file of allRendered) {
        if (seen.has(file.targetPath))
            continue;
        seen.add(file.targetPath);
        const existingEntry = lock.files.find((e) => e.path === file.targetPath);
        if (!existingEntry) {
            items.push({ action: "add", path: file.targetPath, reason: "New file", file });
        }
        else if (existingEntry.status === FileStatus.ConfigInit) {
            items.push({ action: "skip", path: file.targetPath, reason: "Config file (user-owned)" });
        }
        else if (existingEntry.status === FileStatus.Ejected) {
            items.push({ action: "skip", path: file.targetPath, reason: "Ejected by user" });
        }
        else {
            items.push({ action: "update", path: file.targetPath, reason: "Sync from source", file });
        }
    }
    // Check for orphans: files in lock but not in rendered output
    for (const entry of lock.files) {
        if (entry.status === FileStatus.Ejected)
            continue;
        if (!renderedPaths.has(entry.path) && !seen.has(entry.path)) {
            // Source removed upstream
            if (entry.kind === FileKind.Generated) {
                // Auto-delete generated descendants of orphaned sources
                items.push({ action: "delete", path: entry.path, reason: "Source removed upstream (generated auto-delete)" });
            }
            else {
                items.push({ action: "skip", path: entry.path, reason: "Source removed upstream (orphaned)" });
            }
            seen.add(entry.path);
        }
    }
    return {
        items,
        sourceCommit: lock.source.commit,
        sourceRef: options.ref ?? options.version ?? lock.source.ref,
    };
}
// ─── Apply Phase ────────────────────────────────
export async function applySyncPlan(root, lock, plan, _sourceDir) {
    const stagingDir = await createStagingDir(root);
    const actionItems = plan.items.filter((i) => i.action !== "skip");
    const operations = actionItems.map((item, idx) => ({
        id: idx,
        op: (item.action === "delete" ? "delete" : "write"),
        path: item.path,
    }));
    const journal = createJournal(root, "sync", operations, renderedHash(JSON.stringify(lock)));
    await writeJournal(root, journal);
    for (const item of actionItems) {
        if (item.action === "delete") {
            // Backup before delete
            try {
                await createBackup(root, journal.runId, item.path);
                await unlink(join(root, item.path));
            }
            catch {
                // File already gone
            }
            // Remove from lock
            lock.files = lock.files.filter((e) => e.path !== item.path);
        }
        else if (item.file) {
            let content = item.file.content;
            // For source files with slots, merge existing slot content
            if (item.file.transformer === "identity" && item.action === "update") {
                try {
                    const existing = await readFile(join(root, item.path), "utf8");
                    const ext = item.path.match(/\.[^.]+$/)?.[0] ?? ".md";
                    const existingSlots = parseSlots(existing, ext);
                    if (existingSlots.length > 0) {
                        content = mergeSlots(content, existingSlots, ext);
                    }
                }
                catch {
                    // File doesn't exist yet
                }
            }
            // Backup existing file
            if (item.action === "update") {
                try {
                    await access(join(root, item.path));
                    await createBackup(root, journal.runId, item.path);
                }
                catch {
                    // No existing file to backup
                }
            }
            // Stage and commit — write stagedPath to journal for recover --resume
            const stagedPath = await stageFile(stagingDir, content, item.path);
            const journalOp = journal.operations.find((op) => op.path === item.path);
            if (journalOp) {
                journalOp.stagedPath = stagedPath;
                await writeJournal(root, journal);
            }
            await commitStaged(stagedPath, join(root, item.path));
            // Update lock entry
            updateLockEntry(lock, item.file, content);
        }
        await markOperationDone(root, journal, actionItems.indexOf(item));
    }
    // Update lock metadata
    lock.installedAt = new Date().toISOString();
    await writeLock(root, lock);
    await deleteJournal(root);
    await cleanStaging(root);
    return lock;
}
function updateLockEntry(lock, file, content) {
    const existing = lock.files.findIndex((e) => e.path === file.targetPath);
    if (file.transformer === "identity") {
        const ext = file.targetPath.match(/\.[^.]+$/)?.[0] ?? ".md";
        const slots = parseSlots(content, ext);
        const slotBodies = slots.map((s) => s.body);
        const entry = {
            path: file.targetPath,
            kind: FileKind.Source,
            class: FileClass.Core,
            status: FileStatus.Managed,
            sourceCanonicalHash: canonicalHash(file.content, slotBodies),
            localCanonicalHash: canonicalHash(content, slotBodies),
            slotsHash: slotsHash(slotBodies),
        };
        if (existing >= 0)
            lock.files[existing] = entry;
        else
            lock.files.push(entry);
    }
    else {
        // Build source path mapping
        const sourcePathMap = new Map();
        for (const entry of lock.files) {
            if (entry.kind === "source") {
                // Map source-relative to project-relative
                const sourceRel = entry.path.replace(/^\.claude\//, "");
                sourcePathMap.set(sourceRel, entry.path);
                if (entry.path === sourceRel)
                    sourcePathMap.set(entry.path, entry.path);
            }
        }
        const resolvedFromSource = sourcePathMap.get(file.fromSource) ?? file.fromSource;
        const entry = {
            path: file.targetPath,
            kind: FileKind.Generated,
            transformer: file.transformer,
            fromSource: resolvedFromSource,
            renderedHash: renderedHash(content),
            localRenderedHash: renderedHash(content),
            status: FileStatus.Managed,
        };
        if (existing >= 0)
            lock.files[existing] = entry;
        else
            lock.files.push(entry);
    }
}
// ─── CLI Command ────────────────────────────────
export default defineCommand({
    meta: {
        name: "sync",
        description: "Update existing project from source repo",
    },
    args: {
        upgrade: {
            type: "boolean",
            default: false,
            description: "Move to latest version",
        },
        version: {
            type: "string",
            description: "Move to specific version",
        },
        ref: {
            type: "string",
            description: "Branch or SHA",
        },
        "source-config": {
            type: "string",
            description: "Local manifest override",
        },
        "trust-custom": {
            type: "boolean",
            default: false,
            description: "Pre-approve custom transformers",
        },
        "allow-self-sync": {
            type: "boolean",
            default: false,
            description: "Bypass the upstream-repo safety guard (advanced; overwrites local .claude/ with the upstream snapshot)",
        },
    },
    async run({ args }) {
        const root = process.cwd();
        const dryRun = args["dry-run"] === true;
        let lockAcquired = false;
        try {
            if (!dryRun) {
                lockAcquired = await guardMutatingCommand(root, "sync");
            }
            // 1. Read existing lock
            const lock = await readLock(root);
            if (!lock) {
                consola.error("No helpers-lock.json found. Run `helpers init` first.");
                process.exitCode = ExitCode.UsageError;
                return;
            }
            // 1a. Upstream self-sync guard.
            // If cwd contains `helpers.config.ts`, this is the upstream template repo
            // itself. Running `sync` here would overwrite the local `.claude/` tree
            // with the last *published* upstream snapshot — destroying any unpushed
            // authoring work. `helpers regen` is the right command for upstream
            // in-place regeneration. Require explicit `--allow-self-sync` to bypass.
            if (!args["allow-self-sync"]) {
                try {
                    await access(join(root, "helpers.config.ts"));
                    consola.error("Detected `helpers.config.ts` in cwd — this looks like the upstream " +
                        "template repo. Running `sync` here would overwrite your local `.claude/` " +
                        "with the last published upstream snapshot and destroy unpushed changes.\n\n" +
                        "• For upstream maintainers: use `helpers regen` to rebuild generated " +
                        "outputs in-place from the local sources.\n" +
                        "• For consumers that legitimately maintain a mirror: pass `--allow-self-sync` " +
                        "to bypass this guard.");
                    process.exitCode = ExitCode.UsageError;
                    return;
                }
                catch {
                    // No helpers.config.ts — we're in a consumer repo, proceed normally.
                }
            }
            // 2. Determine ref to fetch
            let ref = args.ref ?? args.version;
            if (!ref && !args.upgrade) {
                // Re-validate against pinned version (heal drift only)
                ref = lock.source.ref;
            }
            // 3. Fetch source
            consola.start("Fetching source repo...");
            const { dir: sourceDir, commit } = await fetchSource(lock.source.url, ref, undefined, { offline: args.offline === true });
            // Update lock source info
            if (args.upgrade || args.version || args.ref) {
                lock.source.ref = ref ?? "latest";
                lock.source.commit = commit;
            }
            // 4. Compute plan
            const plan = await computeSyncPlan(lock, sourceDir, lock.targets, {
                upgrade: args.upgrade,
                version: args.version,
                ref: args.ref,
            });
            consola.info(`Sync plan: ${plan.items.filter((i) => i.action === "add").length} add, ${plan.items.filter((i) => i.action === "update").length} update, ${plan.items.filter((i) => i.action === "delete").length} delete, ${plan.items.filter((i) => i.action === "skip").length} skip`);
            // 5. Dry run
            if (dryRun) {
                consola.box("DRY RUN — no files written");
                for (const item of plan.items) {
                    consola.log(`  [${item.action}] ${item.path} — ${item.reason}`);
                }
                return;
            }
            // 6. Apply
            await applySyncPlan(root, lock, plan, sourceDir);
            consola.success("Sync complete.");
        }
        finally {
            await releaseMutatingGuard(root, lockAcquired);
        }
    },
});
// ─── Helpers ────────────────────────────────────
async function discoverSourceFiles(baseDir, patterns) {
    const allFiles = [];
    async function walk(dir) {
        let entries;
        try {
            entries = await readdir(dir, { withFileTypes: true });
        }
        catch (e) {
            consola.warn(`Could not read directory ${dir}:`, e);
            return;
        }
        for (const entry of entries) {
            const fullPath = join(dir, entry.name);
            if (entry.isDirectory()) {
                await walk(fullPath);
            }
            else {
                const relPath = relative(baseDir, fullPath);
                if (patterns.some((p) => matchGlob(p, relPath)))
                    allFiles.push(relPath);
            }
        }
    }
    await walk(baseDir);
    return allFiles.sort();
}
