// clai-helpers/dist/cli/fleet/add-org.js
/**
 * `fleet add-org` command handler.
 * Adds a GitHub org to fleet discovery scope.
 */
import { defineCommand } from "citty";
import consola from "consola";
import { addOrg } from "../../core/fleet/config.js";
import { FleetError } from "../../core/fleet/types.js";
export default defineCommand({
    meta: {
        name: "add-org",
        description: "Add a GitHub org to fleet discovery scope",
    },
    args: {
        org: {
            type: "positional",
            description: "GitHub organization name",
            required: true,
        },
    },
    async run(ctx) {
        const org = ctx.args.org;
        try {
            const config = await import("../../core/fleet/config.js");
            const currentConfig = await config.loadFleetConfig();
            if (currentConfig.scope.orgs.includes(org)) {
                consola.info(`Org '${org}' already in scope.`);
                return;
            }
            await addOrg(org);
            consola.success(`Added '${org}' to fleet scope.`);
        }
        catch (error) {
            if (error instanceof FleetError) {
                consola.error(error.message);
                process.exitCode = 2;
                return;
            }
            throw error;
        }
    },
});
