// clai-helpers/dist/cli/fleet/list.js
/**
 * `fleet list` command handler.
 * Discovers GitHub repos with clai-helpers installed and displays them.
 */
import { defineCommand } from "citty";
import consola from "consola";
import { resolveFleetAuth } from "../../core/fleet/github-api.js";
import { loadFleetConfig } from "../../core/fleet/config.js";
import { discoverFleet } from "../../core/fleet/discovery.js";
import { renderFleetTable, renderFleetJson } from "../../core/fleet/table.js";
import { FleetError } from "../../core/fleet/types.js";
export default defineCommand({
    meta: {
        name: "list",
        description: "List all GitHub repos with clai-helpers installed",
    },
    args: {
        filter: {
            type: "string",
            alias: "f",
            description: "Filter repos by glob pattern",
        },
        json: {
            type: "boolean",
            description: "Output as JSON",
        },
        "no-color": {
            type: "boolean",
            description: "Disable colored output",
        },
        verbose: {
            type: "boolean",
            alias: "v",
            description: "Verbose output",
        },
    },
    async run(ctx) {
        try {
            // 1. Resolve auth
            const auth = resolveFleetAuth();
            // 2. Load config
            const config = await loadFleetConfig();
            // 3. Apply --filter override
            const filterOverride = ctx.args.filter;
            if (filterOverride) {
                config.scope.filter = filterOverride;
            }
            // 4. Run discovery
            const entries = await discoverFleet(config, auth);
            // 5-6. Render output
            if (ctx.args.json) {
                // Structured JSON output to stdout
                console.log(renderFleetJson(entries));
            }
            else {
                const scopeInfo = buildScopeInfo(config);
                console.log(renderFleetTable(entries, {
                    noColor: ctx.args["no-color"],
                    scopeInfo,
                }));
            }
            // 7. Exit 0
            process.exitCode = 0;
        }
        catch (e) {
            if (e instanceof FleetError) {
                switch (e.code) {
                    case "auth/missing":
                        consola.error(e.message);
                        process.exitCode = 2;
                        break;
                    case "github/rate-limited":
                        consola.error(e.message);
                        process.exitCode = 3;
                        break;
                    default:
                        consola.error(e.message);
                        process.exitCode = 2;
                        break;
                }
            }
            else {
                consola.error(e instanceof Error ? e.message : String(e));
                process.exitCode = 1;
            }
        }
    },
});
function buildScopeInfo(config) {
    const parts = [];
    if (config.scope.includeOwnRepos) {
        parts.push("user");
    }
    if (config.scope.orgs.length > 0) {
        parts.push(`${config.scope.orgs.length} orgs (${config.scope.orgs.join(", ")})`);
    }
    return parts.length > 0 ? parts.join(" + ") : "empty scope";
}
