// clai-helpers/dist/cli/fleet/remove-org.js
/**
 * `fleet remove-org` command handler.
 * Removes a GitHub org from fleet discovery scope.
 */
import { defineCommand } from "citty";
import consola from "consola";
import { removeOrg } from "../../core/fleet/config.js";
import { FleetError } from "../../core/fleet/types.js";
export default defineCommand({
    meta: {
        name: "remove-org",
        description: "Remove a GitHub org from fleet discovery scope",
    },
    args: {
        org: {
            type: "positional",
            description: "GitHub organization name",
            required: true,
        },
    },
    async run(ctx) {
        const org = ctx.args.org;
        try {
            const config = await import("../../core/fleet/config.js");
            const currentConfig = await config.loadFleetConfig();
            if (!currentConfig.scope.orgs.includes(org)) {
                consola.info(`Org '${org}' not in scope.`);
                return;
            }
            await removeOrg(org);
            consola.success(`Removed '${org}' from fleet scope.`);
        }
        catch (error) {
            if (error instanceof FleetError) {
                consola.error(error.message);
                process.exitCode = 2;
                return;
            }
            throw error;
        }
    },
});
