// clai-helpers/dist/cli/fleet/sync.js
/**
 * `fleet sync` command handler.
 * Syncs clai-helpers across selected GitHub repos.
 */
import { defineCommand } from "citty";
import consola from "consola";
import { resolveFleetAuth } from "../../core/fleet/github-api.js";
import { loadFleetConfig } from "../../core/fleet/config.js";
import { discoverFleet } from "../../core/fleet/discovery.js";
import { pickEntries, confirmSync } from "../../core/fleet/picker.js";
import { dispatchSync } from "../../core/fleet/modes/index.js";
import { FleetError } from "../../core/fleet/types.js";
import { summarise, exitCode } from "../../core/fleet/types.js";
import { matchGlob } from "../../core/glob.js";
// ── Helpers ──────────────────────────────────────────────────────────
function isInteractiveTty() {
    return process.stdin.isTTY === true;
}
/**
 * Validate that at most one selection flag is present.
 * Returns an error message string if conflict found, null otherwise.
 */
export function validateMutualExclusion(flags) {
    const present = [];
    if (flags.all)
        present.push("--all");
    if (flags.repo && flags.repo.length > 0)
        present.push("--repo");
    if (flags.filter)
        present.push("--filter");
    if (present.length >= 2) {
        return `Cannot use ${present[0]} with ${present[1]}. Pick one selection method.`;
    }
    return null;
}
/**
 * Resolve a non-interactive selection from flags.
 * Must be called AFTER mutual exclusion check passes.
 */
export function selectByFlag(entries, flags) {
    if (flags.all) {
        const active = entries.filter((e) => e.state === "active");
        return { entries: active, source: { kind: "all" } };
    }
    if (flags.repo && flags.repo.length > 0) {
        const repoSet = new Set(flags.repo);
        const matched = entries.filter((e) => repoSet.has(e.fullName));
        // Check for repos not found in fleet
        const matchedNames = new Set(matched.map((e) => e.fullName));
        const missing = flags.repo.filter((r) => !matchedNames.has(r));
        if (missing.length > 0) {
            throw new Error(`Repo '${missing[0]}' not found in fleet.`);
        }
        return { entries: matched, source: { kind: "explicit", repos: flags.repo } };
    }
    if (flags.filter) {
        const matched = entries.filter((e) => matchGlob(flags.filter, e.fullName));
        return { entries: matched, source: { kind: "filter", pattern: flags.filter } };
    }
    return null; // no selection flag → needs interactive picker
}
// ── Command ──────────────────────────────────────────────────────────
export default defineCommand({
    meta: {
        name: "sync",
        description: "Sync clai-helpers across selected GitHub repos",
    },
    args: {
        mode: {
            type: "string",
            description: "Sync mode: pr, push, or patch",
        },
        repo: {
            type: "string",
            description: "Specific repo (owner/repo). Repeatable.",
            repeatable: true,
        },
        filter: {
            type: "string",
            description: "Filter repos by glob pattern",
        },
        all: {
            type: "boolean",
            description: "Sync all discovered repos",
        },
        "patch-output": {
            type: "string",
            description: "Output dir for patch mode",
        },
        yes: {
            type: "boolean",
            alias: "y",
            description: "Skip confirmation prompt",
        },
        "dry-run": {
            type: "boolean",
            description: "Show plan without executing",
        },
        verbose: {
            type: "boolean",
            alias: "v",
        },
        "no-color": {
            type: "boolean",
        },
    },
    async run(ctx) {
        try {
            // 1. Resolve auth
            const auth = resolveFleetAuth();
            // 2. Load config
            const config = await loadFleetConfig();
            // 3. Resolve mode: --mode flag > config.defaultSyncMode > "pr"
            const modeFlag = ctx.args.mode;
            const mode = (modeFlag ?? config.defaultSyncMode);
            if (!["pr", "push", "patch"].includes(mode)) {
                consola.error(`Invalid sync mode: "${mode}". Must be pr, push, or patch.`);
                process.exitCode = 2;
                return;
            }
            // 4. Discover fleet
            consola.start("Discovering fleet...");
            const entries = await discoverFleet(config, auth);
            consola.info(`Discovered ${entries.length} repo(s).`);
            if (entries.length === 0) {
                consola.info("No repos with clai-helpers found.");
                process.exitCode = 0;
                return;
            }
            // 5. Parse selection flags
            const flags = {
                all: ctx.args.all,
                repo: ctx.args.repo,
                filter: ctx.args.filter,
            };
            // 6. Mutual exclusion check
            const conflict = validateMutualExclusion(flags);
            if (conflict) {
                consola.error(conflict);
                process.exitCode = 2;
                return;
            }
            // 7. Resolve selection
            let selection;
            try {
                selection = selectByFlag(entries, flags);
            }
            catch (e) {
                consola.error(e instanceof Error ? e.message : String(e));
                process.exitCode = 2;
                return;
            }
            if (!selection) {
                if (!isInteractiveTty()) {
                    consola.error("Non-interactive terminal. Use --all, --repo, or --filter to select repos.");
                    process.exitCode = 2;
                    return;
                }
                selection = await pickEntries(entries);
            }
            if (selection.entries.length === 0) {
                consola.info("Nothing selected — no repos to sync.");
                process.exitCode = 0;
                return;
            }
            // 8. Dry-run
            if (ctx.args["dry-run"]) {
                consola.box("DRY RUN — no changes will be made");
                for (const entry of selection.entries) {
                    consola.log(`  [${mode}] ${entry.fullName} (${entry.pinnedRef} → ${entry.latestRef})`);
                }
                process.exitCode = 0;
                return;
            }
            // 9. Confirm
            const skipConfirm = ctx.args.yes;
            if (!skipConfirm) {
                if (isInteractiveTty()) {
                    const confirmed = await confirmSync(selection);
                    if (!confirmed) {
                        consola.info("Sync cancelled.");
                        process.exitCode = 0;
                        return;
                    }
                }
            }
            // 10. Sequential sync per entry
            const patchOutputDir = ctx.args["patch-output"] ?? config.patchOutputDir;
            const results = [];
            for (const entry of selection.entries) {
                consola.start(`Syncing ${entry.fullName}...`);
                const result = await dispatchSync(entry, mode, {
                    auth,
                    latestRef: entry.latestRef,
                    patchOutputDir,
                });
                results.push(result);
                switch (result.outcome) {
                    case "succeeded":
                        consola.success(`${entry.fullName}: ${result.outcome} (${result.durationMs}ms)`);
                        break;
                    case "no-op":
                        consola.info(`${entry.fullName}: no-op (already up to date)`);
                        break;
                    case "skipped":
                        consola.warn(`${entry.fullName}: skipped — ${result.reason}`);
                        break;
                    case "failed":
                        consola.error(`${entry.fullName}: failed — ${result.reason}`);
                        break;
                }
            }
            // 11. Build SyncSession, emit summary
            const session = {
                startedAt: new Date().toISOString(),
                endedAt: new Date().toISOString(),
                mode,
                selection,
                results,
                interrupted: false,
            };
            const summary = summarise(session);
            consola.log("");
            consola.info(`Sync complete: ${summary.succeeded} succeeded, ${summary.noOp} no-op, ` +
                `${summary.failed} failed, ${summary.skipped} skipped`);
            // 12. Derive exit code
            process.exitCode = exitCode(session);
        }
        catch (e) {
            if (e instanceof FleetError) {
                switch (e.code) {
                    case "auth/missing":
                        consola.error(e.message);
                        process.exitCode = 2;
                        break;
                    case "github/rate-limited":
                        consola.error(e.message);
                        process.exitCode = 3;
                        break;
                    default:
                        consola.error(e.message);
                        process.exitCode = 2;
                        break;
                }
            }
            else {
                consola.error(e instanceof Error ? e.message : String(e));
                process.exitCode = 1;
            }
        }
    },
});
