// clai-helpers/dist/types/common.js
/**
 * Distinguishes source files (from .claude/) from generated outputs (Copilot/Gemini).
 */
export var FileKind;
(function (FileKind) {
    FileKind["Source"] = "source";
    FileKind["Generated"] = "generated";
})(FileKind || (FileKind = {}));
/**
 * Determines lifecycle behavior during init and sync.
 * - core: Synced on every `sync`. Subject to Protected Slots.
 * - config: Written only on `init`. User owns after first write.
 * - binary: Byte-copied via identity transformer. No slots.
 */
export var FileClass;
(function (FileClass) {
    FileClass["Core"] = "core";
    FileClass["Config"] = "config";
    FileClass["Binary"] = "binary";
})(FileClass || (FileClass = {}));
/**
 * Tracks a file's current management state in the lock file.
 */
export var FileStatus;
(function (FileStatus) {
    FileStatus["Managed"] = "managed";
    FileStatus["ConfigInit"] = "config-init";
    FileStatus["Orphaned"] = "orphaned";
    FileStatus["Ejected"] = "ejected";
})(FileStatus || (FileStatus = {}));
/**
 * CLI exit codes per contracts/cli.md.
 */
export var ExitCode;
(function (ExitCode) {
    ExitCode[ExitCode["Success"] = 0] = "Success";
    ExitCode[ExitCode["UsageError"] = 1] = "UsageError";
    ExitCode[ExitCode["DriftDetected"] = 2] = "DriftDetected";
    ExitCode[ExitCode["StaleJournal"] = 3] = "StaleJournal";
    ExitCode[ExitCode["UntrustedTransformer"] = 4] = "UntrustedTransformer";
    ExitCode[ExitCode["LockSchemaMismatch"] = 5] = "LockSchemaMismatch";
    ExitCode[ExitCode["InternalError"] = 10] = "InternalError";
})(ExitCode || (ExitCode = {}));
