// clai-helpers/dist/core/fleet/modes/patch-mode.js
/**
 * Patch mode: Ephemeral clone → sync → diff → write .patch file.
 * Never mutates GitHub.
 */
import { mkdir, writeFile } from "node:fs/promises";
import { join } from "node:path";
import { FleetError } from "../types.js";
import { createEphemeralClone } from "../ephemeral-clone.js";
import { runSyncPipeline, hasWorkingTreeChanges, getFullDiff } from "./run-sync.js";
/**
 * Sync a repo by generating a .patch file.
 *
 * Flow:
 * 1. Ephemeral clone
 * 2. Run existing sync pipeline against the clone
 * 3. If no changes → no-op
 * 4. Generate diff
 * 5. Write to <patchOutputDir>/<owner>__<repo>.patch
 */
export async function syncPatch(entry, _auth, latestRef, patchOutputDir, _fetchFn) {
    const start = Date.now();
    try {
        // 1. Ephemeral clone
        const { dir, cleanup } = await createEphemeralClone(`https://github.com/${entry.fullName}.git`, entry.defaultBranch);
        try {
            // 2. Run existing sync pipeline against the clone
            const hadChanges = await runSyncPipeline(dir);
            // 3. Check for working tree changes
            const hasChanges = hadChanges || await hasWorkingTreeChanges(dir);
            if (!hasChanges) {
                return {
                    fullName: entry.fullName,
                    outcome: "no-op",
                    mode: "patch",
                    refBefore: entry.pinnedRef,
                    refAfter: entry.pinnedRef,
                    durationMs: Date.now() - start,
                };
            }
            // 4. Generate diff
            const diffContent = await getFullDiff(dir);
            // 5. Write patch file
            const patchFileName = `${entry.fullName.replace("/", "__")}.patch`;
            const patchPath = join(patchOutputDir, patchFileName);
            // Ensure output dir exists
            await mkdir(patchOutputDir, { recursive: true });
            await writeFile(patchPath, diffContent, "utf8");
            return {
                fullName: entry.fullName,
                outcome: "succeeded",
                mode: "patch",
                refBefore: entry.pinnedRef,
                refAfter: latestRef,
                patchPath,
                durationMs: Date.now() - start,
            };
        }
        finally {
            await cleanup();
        }
    }
    catch (error) {
        if (error instanceof FleetError) {
            return {
                fullName: entry.fullName,
                outcome: "failed",
                mode: "patch",
                reason: error.message,
                errorCode: error.code,
                durationMs: Date.now() - start,
            };
        }
        return {
            fullName: entry.fullName,
            outcome: "failed",
            mode: "patch",
            reason: error instanceof Error ? error.message : String(error),
            durationMs: Date.now() - start,
        };
    }
}
