// clai-helpers/dist/core/fleet/modes/pr-mode.js
/**
 * PR mode: Ephemeral clone → sync → branch → commit → push → open PR.
 */
import { execFile } from "node:child_process";
import { promisify } from "node:util";
import { FleetError } from "../types.js";
import { createEphemeralClone } from "../ephemeral-clone.js";
import { findOpenPullRequest, createPullRequest } from "../github-api.js";
import { runSyncPipeline, hasWorkingTreeChanges } from "./run-sync.js";
const execFileAsync = promisify(execFile);
const BRANCH_PREFIX = "clai-helpers-bump";
/**
 * Build a valid git branch name from the latest ref.
 * Falls back to "sync" when the ref is empty/undefined,
 * and strips characters that would make the branch name invalid.
 */
export function buildBranchName(latestRef) {
    const suffix = latestRef && latestRef.trim() ? latestRef.trim() : "sync";
    // Strip trailing slashes and consecutive slashes to keep the name valid
    const sanitized = suffix.replace(/\/+/g, "/").replace(/\/$/, "");
    return `${BRANCH_PREFIX}/${sanitized}`;
}
/**
 * Sync a repo by creating a pull request.
 *
 * Flow:
 * 1. Ephemeral clone
 * 2. Run existing sync pipeline against the clone
 * 3. If no changes → no-op
 * 4. Create branch, commit, push
 * 5. Check for existing open PR (idempotent)
 * 6. Create PR via GitHub API
 */
export async function syncPr(entry, auth, latestRef, fetchFn) {
    const start = Date.now();
    const branchName = buildBranchName(latestRef);
    const resolvedFetch = fetchFn ?? globalThis.fetch;
    try {
        // 1. Ephemeral clone
        const { dir, cleanup } = await createEphemeralClone(`https://github.com/${entry.fullName}.git`, entry.defaultBranch);
        try {
            // 2. Run existing sync pipeline against the clone
            const hadChanges = await runSyncPipeline(dir);
            // 3. Check for working tree changes
            const hasChanges = hadChanges || await hasWorkingTreeChanges(dir);
            if (!hasChanges) {
                return {
                    fullName: entry.fullName,
                    outcome: "no-op",
                    mode: "pr",
                    refBefore: entry.pinnedRef,
                    refAfter: entry.pinnedRef,
                    durationMs: Date.now() - start,
                };
            }
            // 4. Create branch, commit, push
            await execFileAsync("git", ["checkout", "-b", branchName], { cwd: dir });
            await execFileAsync("git", ["add", "-A"], { cwd: dir });
            await execFileAsync("git", ["commit", "-m", `chore(deps): bump clai-helpers to ${latestRef}`], { cwd: dir });
            await execFileAsync("git", ["push", "origin", branchName], { cwd: dir });
            const [owner, repo] = entry.fullName.split("/");
            // 5. Check for existing open PR (idempotent)
            const existingPr = await findOpenPullRequest(owner, repo, branchName, auth, resolvedFetch);
            if (existingPr) {
                return {
                    fullName: entry.fullName,
                    outcome: "succeeded",
                    mode: "pr",
                    refBefore: entry.pinnedRef,
                    refAfter: latestRef,
                    prUrl: existingPr.html_url,
                    durationMs: Date.now() - start,
                };
            }
            // 6. Create PR
            const pr = await createPullRequest(owner, repo, {
                title: `chore(deps): bump clai-helpers to ${latestRef}`,
                body: `Bumps clai-helpers from \`${entry.pinnedRef}\` to \`${latestRef}\`.\n\nAuto-generated by \`clai-helpers fleet sync\`.`,
                head: branchName,
                base: entry.defaultBranch,
            }, auth, resolvedFetch);
            return {
                fullName: entry.fullName,
                outcome: "succeeded",
                mode: "pr",
                refBefore: entry.pinnedRef,
                refAfter: latestRef,
                prUrl: pr.html_url,
                durationMs: Date.now() - start,
            };
        }
        finally {
            await cleanup();
        }
    }
    catch (error) {
        if (error instanceof FleetError) {
            return {
                fullName: entry.fullName,
                outcome: "failed",
                mode: "pr",
                reason: error.message,
                errorCode: error.code,
                durationMs: Date.now() - start,
            };
        }
        return {
            fullName: entry.fullName,
            outcome: "failed",
            mode: "pr",
            reason: error instanceof Error ? error.message : String(error),
            durationMs: Date.now() - start,
        };
    }
}
