// clai-helpers/dist/core/fleet/modes/push-mode.js
/**
 * Push mode: Ephemeral clone → sync → commit → push to default branch.
 */
import { execFile } from "node:child_process";
import { promisify } from "node:util";
import { FleetError } from "../types.js";
import { createEphemeralClone } from "../ephemeral-clone.js";
import { runSyncPipeline, hasWorkingTreeChanges, getHeadSha } from "./run-sync.js";
const execFileAsync = promisify(execFile);
/**
 * Sync a repo by pushing directly to the default branch.
 *
 * Flow:
 * 1. Ephemeral clone
 * 2. Run existing sync pipeline against the clone
 * 3. If no changes → no-op
 * 4. Commit and push to default branch
 * 5. On push rejection, classify error:
 *    - "protected branch" or "policy" → skipped (branch protected)
 *    - "non-fast-forward" or "failed to push some refs" → skipped (push rejected)
 */
export async function syncPush(entry, _auth, latestRef, _fetchFn) {
    const start = Date.now();
    try {
        // 1. Ephemeral clone
        const { dir, cleanup } = await createEphemeralClone(`https://github.com/${entry.fullName}.git`, entry.defaultBranch);
        try {
            // 2. Run existing sync pipeline against the clone
            const hadChanges = await runSyncPipeline(dir);
            // 3. Check for working tree changes
            const hasChanges = hadChanges || await hasWorkingTreeChanges(dir);
            if (!hasChanges) {
                return {
                    fullName: entry.fullName,
                    outcome: "no-op",
                    mode: "push",
                    refBefore: entry.pinnedRef,
                    refAfter: entry.pinnedRef,
                    durationMs: Date.now() - start,
                };
            }
            // 4. Commit and push to default branch
            await execFileAsync("git", ["add", "-A"], { cwd: dir });
            await execFileAsync("git", ["commit", "-m", `chore(deps): bump clai-helpers to ${latestRef}`], { cwd: dir });
            try {
                await execFileAsync("git", ["push", "origin", entry.defaultBranch], { cwd: dir });
            }
            catch (pushError) {
                const stderr = pushError instanceof Error
                    ? pushError.stderr ?? pushError.message
                    : String(pushError);
                const lower = stderr.toLowerCase();
                // Protected branch: contains "protected branch" or "policy"
                if (lower.includes("protected branch") || lower.includes("policy")) {
                    return {
                        fullName: entry.fullName,
                        outcome: "skipped",
                        mode: "push",
                        reason: "git/branch-protected",
                        durationMs: Date.now() - start,
                    };
                }
                // Non-fast-forward: contains "non-fast-forward" or "failed to push some refs"
                if (lower.includes("non-fast-forward") ||
                    lower.includes("failed to push some refs")) {
                    return {
                        fullName: entry.fullName,
                        outcome: "skipped",
                        mode: "push",
                        reason: "git/push-rejected",
                        durationMs: Date.now() - start,
                    };
                }
                // Unknown push error → failed
                throw pushError;
            }
            // Get the pushed SHA
            const pushedSha = await getHeadSha(dir);
            return {
                fullName: entry.fullName,
                outcome: "succeeded",
                mode: "push",
                refBefore: entry.pinnedRef,
                refAfter: latestRef,
                pushedSha,
                durationMs: Date.now() - start,
            };
        }
        finally {
            await cleanup();
        }
    }
    catch (error) {
        if (error instanceof FleetError) {
            return {
                fullName: entry.fullName,
                outcome: "failed",
                mode: "push",
                reason: error.message,
                errorCode: error.code,
                durationMs: Date.now() - start,
            };
        }
        return {
            fullName: entry.fullName,
            outcome: "failed",
            mode: "push",
            reason: error instanceof Error ? error.message : String(error),
            durationMs: Date.now() - start,
        };
    }
}
