// clai-helpers/dist/cli.js
/**
 * CLI entry point using citty.
 * Global flags per contracts/cli.md.
 */
import { defineCommand, runMain } from "citty";
import consola from "consola";
import { ExitCode } from "./types/common.js";
import { readJournal } from "./core/journal.js";
import { acquireProcessLock, releaseProcessLock } from "./core/process-lock.js";
const MUTATING_COMMANDS = new Set([
    "init",
    "sync",
    "regen",
    "remove",
    "remove-target",
    "add-target",
    "eject",
]);
const main = defineCommand({
    meta: {
        name: "helpers",
        version: "0.1.0",
        description: "Transpile .claude/ AI config into Copilot, Gemini, and other formats",
    },
    args: {
        "dry-run": {
            type: "boolean",
            default: false,
            description: "Print plan, write nothing",
        },
        offline: {
            type: "boolean",
            default: false,
            description: "No network. Use giget cache",
        },
        "non-interactive": {
            type: "boolean",
            default: true,
            description: "Never prompt. CI-safe",
        },
        interactive: {
            type: "boolean",
            default: false,
            description: "Prompt on conflicts",
        },
        yes: {
            type: "boolean",
            default: false,
            description: "Auto-confirm destructive operations",
        },
        "no-color": {
            type: "boolean",
            default: false,
            description: "Disable color output",
        },
        json: {
            type: "boolean",
            default: false,
            description: "Output JSON instead of human-readable",
        },
        verbose: {
            type: "boolean",
            default: false,
            description: "Extended logging",
        },
    },
    subCommands: {
        init: () => import("./cli/init.js").then((m) => m.default),
        sync: () => import("./cli/sync.js").then((m) => m.default),
        regen: () => import("./cli/regen.js").then((m) => m.default),
        status: () => import("./cli/status.js").then((m) => m.default),
        diff: () => import("./cli/diff.js").then((m) => m.default),
        eject: () => import("./cli/eject.js").then((m) => m.default),
        remove: () => import("./cli/remove.js").then((m) => m.default),
        "add-target": () => import("./cli/add-target.js").then((m) => m.default),
        "remove-target": () => import("./cli/remove-target.js").then((m) => m.default),
        "list-transformers": () => import("./cli/list-transformers.js").then((m) => m.default),
        doctor: () => import("./cli/doctor.js").then((m) => m.default),
        recover: () => import("./cli/recover.js").then((m) => m.default),
        fleet: () => import("./cli/fleet/index.js").then((m) => m.default),
    },
    setup() {
        // Handle NO_COLOR env var
        if (process.env.NO_COLOR) {
            consola.options.formatOptions = { colors: false };
        }
    },
});
/**
 * Stale journal gate: before executing any mutating command,
 * acquire lock FIRST (atomic), then check for stale journal.
 * Returns true if lock was acquired (caller must release).
 */
export async function guardMutatingCommand(root, commandName) {
    if (!MUTATING_COMMANDS.has(commandName))
        return false;
    // Acquire process lock FIRST — prevents TOCTOU race on journal
    const acquired = await acquireProcessLock(root);
    // Then check for stale journal under the lock
    const journal = await readJournal(root);
    if (journal) {
        const hasIncomplete = journal.operations.some((op) => !op.done);
        if (hasIncomplete) {
            // Release lock before throwing — we're not going to proceed
            if (acquired)
                await releaseProcessLock(root);
            throw new CLIError(`Stale journal detected from a previous "${journal.command}" run (${journal.runId}).\n` +
                `Run \`helpers recover\` to resume, rollback, or abandon before proceeding.`, ExitCode.StaleJournal);
        }
    }
    return acquired;
}
export async function releaseMutatingGuard(root, lockAcquired) {
    if (lockAcquired) {
        await releaseProcessLock(root);
    }
}
/**
 * Custom error with exit code — prevents citty from swallowing our codes.
 */
export class CLIError extends Error {
    code;
    constructor(message, code) {
        super(message);
        this.code = code;
        process.exitCode = code;
    }
}
/**
 * Entry point hook for `bin/helpers.mjs`. Kept as an exported function
 * (not a top-level `runMain(main)` call) so importing this module for its
 * helpers (`guardMutatingCommand`, `CLIError`, etc.) from tests and sibling
 * command files doesn't trigger a CLI parse+exit as a side effect.
 */
export function runCli() {
    runMain(main);
}
