// clai-helpers/dist/core/drift.js
/**
 * Drift detection per FR-018b.
 * Compares local file state against lock file hashes.
 */
import { canonicalHash, slotsHash, renderedHash } from "./hash.js";
import { parseSlots } from "./slots.js";
import { FileStatus } from "../types/common.js";
/**
 * Detect drift for a source entry.
 */
export function detectSourceDrift(entry, localContent, ext) {
    // config-init files never drift
    if (entry.status === FileStatus.ConfigInit) {
        return { isDrift: false, isCustomized: false };
    }
    // Ejected files are not checked
    if (entry.status === FileStatus.Ejected) {
        return { isDrift: false, isCustomized: false };
    }
    const slots = parseSlots(localContent, ext);
    const slotBodies = slots.map((s) => s.body);
    const localCanonical = canonicalHash(localContent, slotBodies);
    // Check if canonical content (outside slots) has changed
    if (localCanonical !== entry.sourceCanonicalHash) {
        // Check if only slots changed
        const localSlots = slotsHash(slotBodies);
        if (localSlots !== entry.slotsHash && localCanonical === entry.localCanonicalHash) {
            // Slots changed but canonical didn't — this is customization, not drift
            return { isDrift: false, isCustomized: true, reason: "Slot content modified" };
        }
        return {
            isDrift: true,
            isCustomized: false,
            reason: "Content modified outside Protected Slots",
        };
    }
    // Check if slots were customized (no drift, but worth noting)
    if (slotBodies.length > 0) {
        const localSlots = slotsHash(slotBodies);
        if (localSlots !== entry.slotsHash) {
            return { isDrift: false, isCustomized: true, reason: "Slot content modified" };
        }
    }
    return { isDrift: false, isCustomized: false };
}
/**
 * Detect drift for a generated entry.
 */
export function detectGeneratedDrift(entry, localContent) {
    const localHash = renderedHash(localContent);
    return localHash !== entry.renderedHash;
}
/**
 * Unified drift detection — dispatches by entry kind.
 */
export function detectDrift(entry, localContent, ext) {
    if (entry.kind === "source") {
        return detectSourceDrift(entry, localContent, ext ?? ".md");
    }
    const isDrift = detectGeneratedDrift(entry, localContent);
    return { isDrift, isCustomized: false, reason: isDrift ? "Generated file modified locally" : undefined };
}
