// clai-helpers/dist/core/fetch.js
/**
 * Source repo fetch wrapper around giget.
 * Auth resolution order per FR-021:
 * 1. Explicit auth param
 * 2. GH_TOKEN env
 * 3. GIGET_AUTH env
 * 4. `gh auth token` subprocess (if gh CLI available)
 */
import { downloadTemplate } from "giget";
import { execSync } from "node:child_process";
import { mkdtemp } from "node:fs/promises";
import { join } from "pathe";
import { tmpdir } from "node:os";
export async function fetchSource(url, ref, auth, options) {
    const resolvedAuth = auth ?? resolveAuth();
    // Resolve `"latest"` (our internal sentinel for "newest release") into an
    // actual git ref. GitHub's tarball endpoint does NOT treat "latest" as a
    // special keyword — passing `#latest` to giget yields /tarball/latest → 404
    // for any repo that doesn't happen to have a tag/branch literally named
    // "latest". We hit the releases API, fall back to the default branch, and
    // only then pass the resolved ref to giget.
    const effectiveRef = ref && ref !== "latest"
        ? ref
        : options?.offline
            ? undefined
            : await resolveLatestRef(url, resolvedAuth);
    const templateUrl = effectiveRef ? `${url}#${effectiveRef}` : url;
    // Download to OS temp dir — NOT cwd — to avoid polluting the project
    const tempDir = await mkdtemp(join(tmpdir(), "helpers-source-"));
    const result = await downloadTemplate(templateUrl, {
        dir: tempDir,
        auth: resolvedAuth,
        offline: options?.offline,
        preferOffline: options?.preferOffline,
        force: true,
        forceClean: true,
    });
    // Resolve commit SHA — skip network call in offline mode
    const commit = options?.offline
        ? (effectiveRef ?? "offline")
        : await resolveCommitSha(url, effectiveRef, resolvedAuth);
    return {
        dir: result.dir,
        commit,
    };
}
/**
 * Resolve the sentinel ref "latest" (or undefined) to a concrete ref.
 *
 * Order:
 *   1. GitHub releases/latest → release `tag_name` (e.g. `v0.2.0`).
 *   2. Repo default_branch (typically `main`).
 *   3. `undefined` (giget will then use the default branch automatically).
 *
 * Returning `undefined` when both lookups fail is safe because giget
 * degrades to the default branch on a bare URL. This keeps fresh projects
 * without any tags working while still preferring tags when they exist.
 */
async function resolveLatestRef(url, auth) {
    const repo = parseGithubRepo(url);
    if (!repo)
        return undefined;
    const headers = {
        Accept: "application/vnd.github+json",
    };
    if (auth)
        headers.Authorization = `Bearer ${auth}`;
    try {
        const releaseRes = await fetch(`https://api.github.com/repos/${repo.owner}/${repo.repo}/releases/latest`, { headers });
        if (releaseRes.ok) {
            const release = (await releaseRes.json());
            if (release.tag_name)
                return release.tag_name;
        }
    }
    catch {
        // Network glitch — fall through to default_branch lookup
    }
    try {
        const repoRes = await fetch(`https://api.github.com/repos/${repo.owner}/${repo.repo}`, { headers });
        if (repoRes.ok) {
            const data = (await repoRes.json());
            if (data.default_branch)
                return data.default_branch;
        }
    }
    catch {
        // Ignore — fall through to undefined
    }
    return undefined;
}
function parseGithubRepo(url) {
    const match = url.match(/github[.:]([^/]+)\/([^/#]+)/);
    if (!match)
        return null;
    return { owner: match[1], repo: match[2].replace(/\.git$/, "") };
}
export function resolveAuth() {
    // 1. GH_TOKEN env
    if (process.env.GH_TOKEN)
        return process.env.GH_TOKEN;
    // 2. GIGET_AUTH env (giget reads this natively, but we're explicit)
    if (process.env.GIGET_AUTH)
        return process.env.GIGET_AUTH;
    // 3. gh auth token subprocess
    try {
        const token = execSync("gh auth token", {
            encoding: "utf8",
            timeout: 5000,
            stdio: ["pipe", "pipe", "pipe"],
        }).trim();
        if (token)
            return token;
    }
    catch {
        // gh not available or not authenticated
    }
    return undefined;
}
/**
 * Resolve commit SHA via GitHub API or fall back to ref as-is.
 * giget downloads flat snapshots without .git/, so we can't use git rev-parse.
 */
async function resolveCommitSha(url, ref, auth) {
    // Parse GitHub owner/repo from URL
    const match = url.match(/github[.:]([^/]+)\/([^/#]+)/);
    if (!match)
        return ref ?? "unknown";
    const [, owner, repo] = match;
    const branch = ref ?? "HEAD";
    try {
        const headers = {
            Accept: "application/vnd.github+json",
        };
        if (auth)
            headers.Authorization = `Bearer ${auth}`;
        const response = await fetch(`https://api.github.com/repos/${owner}/${repo}/commits/${branch}`, { headers });
        if (response.ok) {
            const data = (await response.json());
            return data.sha;
        }
    }
    catch {
        // Network error — return ref as-is
    }
    return ref ?? "unknown";
}
