// clai-helpers/dist/core/manifest.js
/**
 * Manifest loader via c12 per contracts/manifest.md.
 * Loads helpers.config.ts from the source repo.
 */
import { loadConfig } from "c12";
/**
 * Identity helper for type-safe manifest authoring.
 * Usage: `export default defineHelpersConfig({ ... })`
 */
export function defineHelpersConfig(config) {
    return config;
}
/**
 * Load and validate manifest from source directory.
 */
export async function loadManifest(sourceDir, overridePath) {
    const { config } = await loadConfig({
        cwd: sourceDir,
        name: "helpers",
        defaults: { version: 1, sources: [], targets: {} },
        ...(overridePath ? { overrides: { configFile: overridePath } } : {}),
    });
    const manifest = config;
    validateManifest(manifest);
    return manifest;
}
function validateManifest(manifest) {
    if (manifest.version !== 1) {
        throw new Error(`Manifest version ${manifest.version} is not supported. Expected 1.`);
    }
    if (!manifest.sources || manifest.sources.length === 0) {
        throw new Error("Manifest must have at least one source pattern.");
    }
    const targetNames = Object.keys(manifest.targets);
    if (targetNames.length === 0) {
        throw new Error("Manifest must have at least one target.");
    }
    // Check for duplicate output paths across all targets
    const outputPaths = new Set();
    for (const [targetName, target] of Object.entries(manifest.targets)) {
        for (const pipeline of target.pipelines) {
            // Validate match is subset of sources (simple check: at least one source pattern could match)
            const matchesAny = manifest.sources.some((src) => patternsOverlap(src, pipeline.match));
            if (!matchesAny) {
                throw new Error(`Pipeline match "${pipeline.match}" in target "${targetName}" is not covered by any source pattern.`);
            }
            // Check output template variables
            const validVars = /\{\{(name|relativePath|subpath|ext)\}\}/g;
            const stripped = pipeline.output.replace(validVars, "");
            const invalidVars = stripped.match(/\{\{(\w+)\}\}/);
            if (invalidVars) {
                throw new Error(`Invalid template variable "{{${invalidVars[1]}}}" in output "${pipeline.output}" of target "${targetName}".`);
            }
            // Only flag duplicates for literal output paths (no template variables).
            // Templates with {{name}}, {{relativePath}}, etc. produce different paths per source file.
            const hasTemplateVars = /\{\{(name|relativePath|subpath|ext)\}\}/.test(pipeline.output);
            if (!hasTemplateVars) {
                if (outputPaths.has(pipeline.output)) {
                    throw new Error(`Duplicate output path: "${pipeline.output}" across targets.`);
                }
                outputPaths.add(pipeline.output);
            }
        }
    }
}
/**
 * Simple heuristic: two glob patterns might overlap if they share a common base.
 * This is not a full glob intersection check — it catches obvious mismatches.
 */
function patternsOverlap(source, match) {
    // "**/*" matches everything
    if (source === "**/*" || match === "**/*")
        return true;
    if (source === match)
        return true;
    // Check if the directory prefix matches
    const sourceBase = source.split("/")[0];
    const matchBase = match.split("/")[0];
    if (sourceBase === "**" || matchBase === "**")
        return true;
    if (sourceBase === matchBase)
        return true;
    // Single file patterns
    if (!source.includes("*") && !match.includes("*")) {
        return source === match;
    }
    return false;
}
