// clai-helpers/dist/core/slots.js
/**
 * Protected Slots parser & merger.
 * Format-aware marker detection per FR-013.
 */
const MARKER_PAIRS = {
    // Markdown / HTML
    md: {
        start: /^<!--\s*HELPERS:CUSTOM\s+START\s*-->$/,
        end: /^<!--\s*HELPERS:CUSTOM\s+END\s*-->$/,
        rawStart: "<!-- HELPERS:CUSTOM START -->",
        rawEnd: "<!-- HELPERS:CUSTOM END -->",
    },
    // YAML / hash-comment languages
    yaml: {
        start: /^#\s*HELPERS:CUSTOM\s+START$/,
        end: /^#\s*HELPERS:CUSTOM\s+END$/,
        rawStart: "# HELPERS:CUSTOM START",
        rawEnd: "# HELPERS:CUSTOM END",
    },
    // JS / TS / JSONC
    js: {
        start: /^\/\/\s*HELPERS:CUSTOM\s+START$/,
        end: /^\/\/\s*HELPERS:CUSTOM\s+END$/,
        rawStart: "// HELPERS:CUSTOM START",
        rawEnd: "// HELPERS:CUSTOM END",
    },
    // TOML
    toml: {
        start: /^#\s*HELPERS:CUSTOM\s+START$/,
        end: /^#\s*HELPERS:CUSTOM\s+END$/,
        rawStart: "# HELPERS:CUSTOM START",
        rawEnd: "# HELPERS:CUSTOM END",
    },
};
function getMarkerPair(ext) {
    const normalized = ext.toLowerCase().replace(/^\./, "");
    if (normalized === "json")
        return null; // FR-013: plain JSON rejection
    if (["md", "html", "htm", "xml", "svg"].includes(normalized)) {
        return MARKER_PAIRS.md;
    }
    if (["yaml", "yml", "sh", "bash", "zsh", "py", "rb", "toml", "conf", "ini", "cfg"].includes(normalized)) {
        return MARKER_PAIRS.yaml;
    }
    if (["js", "ts", "mjs", "mts", "cjs", "cts", "jsx", "tsx", "jsonc"].includes(normalized)) {
        return MARKER_PAIRS.js;
    }
    // Default to markdown-style for unknown extensions
    return MARKER_PAIRS.md;
}
/**
 * Parse Protected Slots from file content.
 * Returns null-safe marker pair for plain JSON (rejected per FR-013).
 */
export function parseSlots(content, ext) {
    const pair = getMarkerPair(ext);
    if (!pair)
        return []; // JSON — no slots
    const lines = content.split("\n");
    const slots = [];
    let inSlot = false;
    let currentStartLine = -1;
    let currentBody = [];
    let currentStartMarker = "";
    for (let i = 0; i < lines.length; i++) {
        const trimmed = lines[i].trim();
        if (!inSlot && pair.start.test(trimmed)) {
            inSlot = true;
            currentStartLine = i;
            currentStartMarker = lines[i];
            currentBody = [];
        }
        else if (inSlot && pair.end.test(trimmed)) {
            slots.push({
                startMarker: currentStartMarker,
                endMarker: lines[i],
                body: currentBody.join("\n"),
                startLine: currentStartLine,
                endLine: i,
            });
            inSlot = false;
            currentBody = [];
        }
        else if (inSlot) {
            currentBody.push(lines[i]);
        }
    }
    return slots;
}
/**
 * Validate that all slot markers are properly paired.
 * Returns an Error if unbalanced, null if valid.
 */
export function validateSlotPairing(content, ext) {
    const pair = getMarkerPair(ext);
    if (!pair)
        return null; // JSON — no slots possible
    const lines = content.split("\n");
    let depth = 0;
    for (let i = 0; i < lines.length; i++) {
        const trimmed = lines[i].trim();
        if (pair.start.test(trimmed)) {
            depth++;
            if (depth > 1) {
                return new Error(`Nested slot marker at line ${i + 1}. Nesting is not supported.`);
            }
        }
        else if (pair.end.test(trimmed)) {
            depth--;
            if (depth < 0) {
                return new Error(`Unmatched slot END marker at line ${i + 1} without a preceding START.`);
            }
        }
    }
    if (depth > 0) {
        return new Error("Unmatched slot START marker without a closing END marker.");
    }
    return null;
}
/**
 * Merge new upstream content with existing slot bodies.
 * Replaces slot bodies in newContent with corresponding bodies from existingSlots.
 */
export function mergeSlots(newContent, existingSlots, ext) {
    if (existingSlots.length === 0)
        return newContent;
    const newSlots = parseSlots(newContent, ext);
    if (newSlots.length === 0)
        return newContent;
    const lines = newContent.split("\n");
    const result = [];
    let slotIdx = 0;
    for (let i = 0; i < lines.length; i++) {
        const currentSlot = newSlots[slotIdx];
        if (currentSlot && i === currentSlot.startLine) {
            // Write start marker
            result.push(lines[i]);
            // Write existing slot body (preserved content)
            const existingSlot = existingSlots[slotIdx];
            if (existingSlot) {
                result.push(existingSlot.body);
            }
            else {
                // No matching existing slot — keep new content
                result.push(currentSlot.body);
            }
            // Skip to end marker
            i = currentSlot.endLine;
            result.push(lines[i]);
            slotIdx++;
        }
        else {
            result.push(lines[i]);
        }
    }
    return result.join("\n");
}
